Revise AI policy (#2551)
bevyengine/bevy-website: content/learn/contribute/policies/ai.md

weight = 1
+++

> Bevy, as a project and a community, values:

Author: **SENTIMENT THREAD ON REVISED POLICY**

Reviewer: Much better; this version addresses the concerns I had with the first proposal.

Reviewer 2: It is much better, although doesn't go far enough to establish that the "goal" is to not use AI. However, this policy I can live with

Reviewer 3: As someone who hasn't been involved in previous discussion, and has fairly mixed feelings about what the policy allows, I think the revisions are good. Much clearer and more direct.

I do have one major issue with the "we are wary" section near the beginning. I've written more here: https://github.com/bevyengine/bevy-website/pull/2551#discussion_r3740622183

Author: @jasonaowen said (in a different thread, moved to be tidy):

> I fear that this policy change will drive away people that would be good for the Bevy community, both active contributors like those who have commented on this PR and users like myself that want nothing to do with AI, without resolving the issues that led to this proposal.
> 
> The ongoing harms that bad-faith contributors are causing are terrible, and preventing them and protecting the project and the maintainers is important. However, I don't think this policy will accomplish that.
> 
> I'd like to suggest a third option, instead of maintaining the status quo or adopting this policy change: moving to Codeberg. Every page of this repo - every page on this website - includes at least one ad for GitHub's AI assistant, Copilot. Moving to an environment filled with people who already share our values should relieve at least some of the pressure. I know how much work would be involved, so I do not make this suggestion lightly.
> 
> One of the reasons I started using Bevy is because of its strong stance against AI. I felt like I had found a community who shared my values and recognized the harm the AI industry was inflicting, and made the ethical choice. Turning away from that choice breaks my heart.
> 
> This isn't what I wanted my first contribution to Bevy to be, and I hope it won't be my last. Thank you for building an amazing community and technology.

> 1. **Community:** Collaboration and camaraderie between contributors

Reviewer 3: ```suggestion
The Bevy project values:

1. **Community:** Collaboration and camaraderie between contributors
```

Style suggestion - I think the previous version awkwardly scanned as "Bevy, as a ... community, values ... community".

2. **Growth:** Developing understanding and wisdom within our contributors
3. **Quality:** Creating a reliable, high-quality foundation for others to build on

We are wary of the use of AI in the development of Bevy,
and are particularly troubled by its impact on artists,

> but tolerate careful human-driven AI use in some areas consistent with the above values.

Reviewer 4: I think we should reword this to "but tolerate careful human-driven AI use in some technical, non-art areas consistent with the above values."

I don't want it to sound like we are interested in using "careful human-driven AI use" for art. That would definitely turn a lot of artists away.

Reviewer 5: The issue here is that a strong point of contention with the previous version of the policy was that some contributors consider code itself as an art form, so the policy was updated to explicitly avoid making that value judgement/classification.

Reviewer 4: Ok. Then how about "but tolerate careful human-driven AI use in some coding-related tasks consistent with the above values."?

Reviewer 3: My two cents on the artist impact part is that it's ambiguous and counter-productive. I'm going to be a bit negative here as I think the rest of the policy is good, but this section lets it down. I'm speaking as a coder who has worked professionally supporting artists in game dev, and had conversations with artists around generative AI and other technologies that threaten jobs. 

First, the word "artists" is context dependent, and in game dev is sometimes used quite narrowly. If I asked people I've worked with "are you an artist?", many would reply "no, I'm a writer/designer/animator/etc". And if I said "but your work is art?" they'd probably say "oh yes, I'm an artist, but I'm not an _artist_ artist". Some of these disciplines are in just as precarious a position as 2D/3D visual artists. Does the statement include them? I can't tell.

Either way, it feels weird to pick out some disciplines, and begs to be misread as a knock on other disciplines. And it has a slight whiff of corporate platitude - the kind of "our number 1 priority is our employees" statement that precedes the complete opposite. Which is a pity as it's actually followed by a clear "no AI media" rule that will be far more meaningful and reassuring!

I would try something more direct:

```diff
 3. **Quality:** Creating a reliable, high-quality foundation for others to build on
 
-We are wary of the use of AI in the development of Bevy,
-and are particularly troubled by its impact on artists,
-but tolerate careful human-driven AI use in some areas consistent with the above values. 
-
-In concrete terms, Bevy expects its community members to uphold the following norms around AI:
+To uphold these values, we limit the use of generative AI within the Bevy project. This means:
 
 1. **No AI-generated media.** Images, [etc]
```

I think this is clear that we are limiting the use of AI _because_ of our values, and then gets straight onto what that means in practice. Some might say that it buries the lede, since the new policy is more permissive than the old. But I think it honestly reflects that the new policy is far more restrictive than it is permissive.

Reviewer 4: Yeah, I really like greeble-dev's suggestion!

Author: Added a "code-related" qualifier.

In concrete terms, Bevy expects its community members to uphold the following norms around AI:

> 1. **No AI-generated media.** Images, 3d models, audio assets (and so on) submitted to Bevy may not be authored or modified with the use of generative AI tools. Use a permissively licensed asset, make one by hand, or commission an artist.

Reviewer 6: term 1.

Reviewer 6: Regarding "permissively licensed": I think this should, in the future/once it exists, link to a guideline/policy/explanation about the asset licenses bevy accepts.

> 2. **No AI-generated prose.** AI cannot be used to write public-facing prose, including documentation, issues, PR descriptions, or release notes.

Reviewer 7: Nitpick: "must not" instead of "cannot".

Reviewer 7: Or better, phrased actively: "Do not use AI to write"...

> 3. **No AI-generated communication.** We want to talk to *you*, not a machine. Do not paste AI-generated output directly into conversations with humans, no matter where they take place. AI may be used to assist for translation and accessibility purposes, but please include the original untranslated text as well.

Reviewer 6: term 3.

Reviewer 6: ```suggestion
3. **No AI-generated communication.** We want to talk to *you*, not a machine. Do not paste AI-generated output directly into conversations with humans, no matter where they take place. AI may be used to assist with translation and accessibility purposes, but please include the original untranslated text as well.
```

Reviewer 6: I once again want to recommend, for translation, mentioning specific tools, to guide people away from using general purpose LLMs for that. My personal recommendation is DeepL as it keeps the tone, exact wording, etc. intact far better than what i've seen of people asking a general purpose LLM to translate.

Reviewer 6: ```suggestion
3. **No AI-generated communication.** We want to talk to *you*, not a machine. Do not paste AI-generated output directly into conversations with humans, no matter where they take place. AI may be used to assist for translation and accessibility purposes, but please include the original untranslated text as well, for example as a [collapsed section][collapsed section] on github.
```

If not, the [collapsed section] link below should probably be removed, since its not used anywhere.

> 4. **No AI-authored commits.** This includes "Co-authored by" and is enforced via automated checks in CI.

Reviewer 3: term 4

> **No AI-authored commits.** This includes "Co-authored by" and is enforced via automated checks in CI.

Reviewer 3: I was initially confused by this point, even though I'm familiar with the previous situations around git "Co-authored by" trailers. I suspect that someone unfamiliar with that stuff might misread it completely - in particular, "No AI-authored commits" could be misread as "No AI-authored _code_".

Maybe this whole point should be moved to a separate "practical considerations for PRs" section, and expanded to be explicit that it's about git trailers? That section could also include advice around translation (i.e. including original text in collapsed sections), examples of how to disclose AI use, and other things that are more prescriptive and good practice related.

Reviewer 6: > "No AI-authored commits" could be misread as "No AI-authored code".

term 5 means that fully/only AI-authored code isn't really welcome anyways, as that means term 5 isn't being followed, so this being a potential misunderstanding does not seem like a big issue to me

Reviewer 8: Perhaps “No AI attribution in commit messages” for sharper language?

Reviewer 6: > No AI attribution in commit messages

I critized wording like this in the previous draft for pushing towards just stripping the attribution, while still submitting wholly AI authored commit, which obviously shouldn't happen under term 5. I would ask that, if this is changed, the wording is chosen such that it can't be interpreted as "just remove AI attribution from the fully AI commit"

Reviewer 5: ```suggestion
4. **No AI-authored commits.** This includes, _but is not limited to_, `Co-Authored-By` commits (rejected via automated checks in CI). We view commit messages as prose/human communication, so items 2 and 3 apply here as well.
```

Reviewer 3: > term 5 means that fully/only AI-authored code isn't really welcome anyways, as that means term 5 isn't being followed, so this being a potential misunderstanding does not seem like a big issue to me

My concern is that the phrasing _invites_ misunderstanding. To correctly interpret the current phrasing, the reader has to make the connection that "AI-authored commits" and "co-authored by" refers _specifically_ to git trailers. I think a lot of people will not make that connection, or even know what a git trailer is.

Blender's equivalent is:

> For Git commits, only humans can be commit authors or listed as  `Co-authored-by`.

Putting aside whether the policy itself is a good idea, I think Blender's is harder to misunderstand because it starts by referencing Git.

Reviewer 2: I found it odd to ban Co-authored-by commits. It somewhat goes against the point of requiring disclosure. Likewise those commit messages are an important signal for me when I am looking for projects to avoid, or if I'm confused by a change and looking to see if it were ai generated.

Reviewer 5: The key issue is the Co-Authored-By tag is meant for _human_ attribution. It shows up in places like blame etc. AI models (at least for the foreseeable future) cannot answer for or be responsible their actions/decisions, so it's arguable that this is misuse by the labs. Linux actually uses a different tag: `Assisted-by` which I feel like it's more appropriate, but I believe tooling support is not there yet, and it's probably harder to get people on board with using that

https://github.com/torvalds/linux/blob/master/Documentation/process/coding-assistants.rst

> 5. **Ownership.** As a contributor, you are personally responsible for everything you submit. You must personally understand every line of code or documentation that you put forward for review, and be able to articulate the design rationale and implementation tradeoffs when asked. Work within your skill level and knowledge of the project.

Reviewer 4: I think we should also add something along the lines of "Just asking the LLM why it made a design decision and parroting back what it says does NOT count as understanding, even if you rewrite what it said in your own words." This *is* kind of implied by the other points, but I think it would do good to make this very clear.

Reviewer 5: I like the sentiment/idea but that wording is too long IMO, we need to make it shorter/simpler to match the terseness of the rest of the policy

Reviewer 4: Fair. Perhaps we could just change "You must personally understand" to "You (not the AI) must personally understand"?

Author: I can't find a wording for this that I'm happy with. "You must personally understand" already captures this idea IMO (that was certainly my intent when drafting!), and I don't think that making the wording worse but more explicit will improve matters.

Reviewer 9: "As a human author, you are expected to understand every line of code..."

> 5. **Ownership.** As a contributor, you are personally responsible for everything you submit. You must personally understand every line of code or documentation that you put forward for review, and be able to articulate the design rationale and implementation tradeoffs when asked. Work within your skill level and knowledge of the project.

Reviewer 7: Would say "You must understand" as there is not another way to understand aside from personally.

Author: I like the flow better like this.

6. **Disclosure.** Any use of AI to contribute to Bevy must be disclosed, with context about how it was used.

> 7. **Quality.** Bevy holds a high bar for quality, and is willing to take the time to do things well. AI-assisted PRs will be held to a rigorous standard of review: correctness, tests, rationale, incremental reviewable units, appropriate design and consensus building and so on. 

Reviewer 6: term 7.

Reviewer 6: For "enforcing" this, i would argue for a new issue label to mark issues for more in-depth review that are suspected to be somewhat lacking in quality *for any reason*. Something that isn't specific to AI removes the issues with "witch-hunting" and could honestly be useful even for fully human-written PRs. 

Ideally the phrasing of the label should be neutral, maybe result-oriented (not just `low-quality`), maybe `review-deeply` or `ensure-quality`? (eugh, i don't really like either, maybe something with `nominated-for-` but i have no clue how to concices express the idea of "a triage team member recommends this issue gets looked at more closely")

Reviewer 5: Maybe something like `S-Needs-Refinement`?

Author: We have `S-Needs-Help` already, which I think is a good fit to start? Very happy to just add more without any ceremony as we see how this plays out.

> 7. **Quality.** Bevy holds a high bar for quality, and is willing to take the time to do things well. AI-assisted PRs will be held to a rigorous standard of review: correctness, tests, rationale, incremental reviewable units, appropriate design and consensus building and so on. 

Reviewer 7: Suggestion: "AI-assisted PRs will be held to a rigorous standard of review, including correctness, tests, rationale, incremental reviewable units, appropriate design and consensus building."

Author: I like the original wording slightly better!

> 8. **Harassment.** Do not harass, demean, pressure, or bully others. This continues to apply, even when the basis of your disagreement centers around use of, abstinence from or views on AI. See the [appendix on harassment](#appendix-harassment-policy-details) for examples of acceptable and unacceptable behavior.

Reviewer 6: term 8.

Reviewer 6: (also re. Appendix: Harassment Policy Details): The main thing i'm not sure about, because its something i would like to be able to do, is whether this forbids me from recommending people away from relying on AI. Would that fall under pressuring?

Its not clear from the Appendix either: If its a reply to a specific person, i'm thinking someone in the discord help forum asking something along the lines of "<insert ai model> told me X but it isn't working", i'd like to know if something along the lines of:
> Heres the relevant documentation/<explanation of the thing>. I recommend not relying on <AI/LLMs/specific tool name>, instead reading and figuring things out yourself.

would fall under term 8. or generally be considered fine.

## Appendix: Harassment Policy Details

We understand that AI is a controversial topic, with important societal implications, many of which directly impact topics the Bevy community cares about.
Nevertheless, Bevy's [Code of Conduct] continues to apply when AI-related topics are discussed.
To help you understand where Bevy draws the line, the following AI-related behaviors are not acceptable anywhere within the Bevy community:

- inflammatory comments about AI or patterns of behavior that seek to get a rise out of others
- comments that imply that AI use is or will be mandatory (inside of Bevy or outside of it), or which otherwise disparage those who do not want to use it / think it should be used
- refusal to disengage from an argument with a specific individual after being asked to stop
- unsubstantive negative comments or reactions in response to contributors or community members who disclose AI use

If you have frustrations or concerns with other community members' behavior or project-wide rules, come to the moderation team.

> We will try our best to resolve problems, even if it is a repeated pattern of subthreshold behavior!

Reviewer 10: I've been thinking a bit about AI Boosterism with regards to 'a repeated pattern of subthreshold behavior'.

The most blunt rhetoric an AI Salesperson could use is 'adopt or you'll be left behind'. This is very unhelpful advice and already prohibited, so I won't dwell on it. There are, however, subtler things one can do to 'champion' AI, which aren't explicitely prohibited, but still degrade the experience of a community space: Nobody enjoys being marketed to, especially if you've already decided you dislike the product.

If you want a quick impression of how this works, anthropic made this [black hole of corporate soullessness](https://code.claude.com/docs/en/champion-kit).

I don't think anyone is literally following this playbook, nor that the enthusiasm felt for these agents is insincere, but I do see 'repeated patterns of subthreshold behavior' that I would qualify as AI marketing.

The practical solution is to more strictly silo the AI conversation away into `#machine_learning`. You can talk about the benefits of your multi-agent workflow all day, but I don't have to see it. In some sense, this is similar to the dedicated `#code_editor` channels of old, where you would go to fight the vim-emacs war, only with slightly lower stakes.

Note that I would only like to silo away (subtle) AI Marketing, the zeitgeist being what it is and Bevy (rightly imo) allowing the discussion of politics, the topic of AI *will* come up. The best rule I can come up with would be something along the lines of:

```
...not acceptable within the Bevy community:
- insistent marketing for AI products or the concept of AI. Keep the discussion of AI workflows within the dedicated `#machine-learning` channel.
```

Author: Mmm, I like the idea of something like this a lot! This is unwelcome behavior that can be quite frustrating, and a good bright-line example of "what does boosterism mean" that is focused on social conduct, not beliefs.

I'll chew on this and write this out proper tomorrow.

Reviewer 5: I would even suggest splitting #machine-learning into two channels, one for discussing articles, papers, news, techniques, inference frameworks etc and one for discussing/sharing/hyping AI coding stuff specifically

The rationale being that you might be interested in keeping tabs or chatting about it in a more serious fashion without spam, kinda like off-topic vs memes

If you are new to the Bevy community, you might ask: why not simply ban AI contributions and be done with it?

In brief: [we tried that](https://github.com/bevyengine/bevy-website/pull/2204).
Motivated by concerns around quality, ethics and legal risk, the Bevy maintainers voted to ban all AI-generated contributions.
It didn't go very well.

In practice, we found that we could not fairly and reliably enforce or adjudicate this policy.
AI-generated prose is *sometimes* detectable, but rarely to the level of evidence desirable.
AI-generated code is even less so.
AI-generated bug detection, research, or design input could be *completely* invisible,
and while this was not barred, many community members were unhappy with such use.

Submissions were made that were *maybe* AI-generated with no way to ever prove it one way or another,
forcing reviewers and moderators to constantly decide to act on or overlook violations on the slimmest of evidence.
Any accusation threatened complete rejection of submitted work regardless of quality or effort,
with a looming possibility of formal and informal exclusion from the community.

> Reviwers were caught between unproven, disruptive suspicion and blithe naïveté,

Reviewer 11: ```suggestion
Reviewers were caught between unproven, disruptive suspicion and blithe naïveté,
```

7. **Design:** Writing is thinking.
  While AI can be helpful during the design process, it should be used to explore the space, not to simply generate a solution.
  Publicly shared design documents (or problem definitions) must be personally authored, synthesizing your own understanding of the problem space and any proposed solution.
  We want to foster genuine expertise, and feel that taking the time to draft plans personally is critical for both learning and refinement. Care and effort spent planning pays dividends later.
8. **Art:** Out of respect for the craft of artists, AI must not be used for creating or manipulating art assets: find a permissively licensed asset, commission something appropriate, or make one by hand.
9. **Harassment:** Do not harass, pressure, or demean others because of their AI use or lack thereof.
  See the attached Appendix for examples of acceptable and unacceptable behavior.
  There are a wide range of perspectives within our community, and community health is more important than winning arguments on the internet or accomplishing any particular bit of work.

> 10. **Refusal:** Contributors are allowed to refuse to review AI-assisted code for any reason, or reject AI-assisted reviews. Please, do not pressure them to change their minds.

Reviewer 12: This point here highlights two problems when compared to the rest of the document:

1. how contributors could be pushed out for a simple reason: Contributors that do less aren't as valued as contributors that do more. A refusal is not enough to prevent this from creating a situation where contributors who won't want to review AI-assisted PRs, maybe even from other maintainers, will find themselves getting sidelined over time. One contributor that doesn't want to review could have another step in to review in their place. Refusals without closing create such a situation, even if the intent here is "don't burn out contributors/maintainers". But if there is no outright ban on AI use, contributors can't close a PR they don't like, and thus will be pressured to review it anyway.

2. If PRs that are visibly or disclosed to be AI assisted get refused/rejected more often than other PRs, then the AI usage will get hidden anyway, and thus you are back in a situation where you are double guessing whether a new contributor is using AI or not.

Bad actors don't particularly *care* about following a policy, as far as they can get code committed and thus build their own reputation (at the expense of other projects). And we've also created a catch-22 situation where by giving a little wiggle room to AI contributions, you create a potentially toxic situation anyway for contributors where the only way to "win" and not be forced out is to use AI or accept AI.

Reviewer 5: I think under such a scheme only SMEs would be "forced" to review AI code, or at least for every subject matter that has a non-AI-reviewing SME we'll need to appoint an AI-reviewing SME otherwise all AI PRs for that area get effectively blocked.

I don't think we are more or less valued for the volume of our work. Barring the few people under the Bevy Foundation payroll we're volunteers doing this on our free time because we enjoy it, so any sense of value is self-ascribed.

Re: reputation and other external benefits of contributing to open source, (e.g. employability) if anything LLMs have already mostly or fully eroded this, sadly, so I don't think we'd be significantly hindering anything.

Reviewer 12: > I think under such a scheme only SMEs would be "forced" to review AI code, or at least for every subject matter that has a non-AI-reviewing SME we'll need to appoint an AI-reviewing SME otherwise all AI PRs for that area get effectively blocked.

And thus, the non-AI reviewing SME could get sidelined with enough refusals over time. Why would someone who is anti-AI want to become an SME in this environment?

> I don't think we are more or less valued for the volume of our work.

That has not been my experience, unfortunately.

> Re: reputation and other external benefits of contributing to open source, (e.g. employability) if anything LLMs have already mostly or fully eroded this, sadly, so I don't think we'd be significantly hindering anything.

This was about describing motivations of bad actors. What I am saying here is that attempts to draft a policy to "encourage" good behaviour out of bad actors is futile.

Reviewer 5: > And thus, the non-AI reviewing SME could get sidelined with enough refusals over time. Why would someone who is anti-AI want to become an SME in this environment?

I see what you mean now, though this will only become an issue if AI contributions reach a significant threshold. SMEs don't only do reviews so even a theoretical scenario where the vast majority of PRs are AI, an SME "duo" could still have one tasked with reviewing and other tasked with other SME responsibilities

> What I am saying here is that attempts to draft a policy to "encourage" good behaviour out of bad actors is futile.

Agreed, the best we can do is not encourage bad behavior, as we can't even properly discourage it

Reviewer 12: > I see what you mean now, though this will only become an issue if AI contributions reach a significant threshold. SMEs don't only do reviews so even a theoretical scenario where the vast majority of PRs are AI, an SME "duo" could still have one tasked with reviewing and other tasked with other SME responsibilities

I would argue that threshold has been reached already, given that this policy revision is being drafted at all.

And another case to point out: Someone who wants to contribute more, but doesn't do as much review (perhaps not wanting to deal with AI-assisted/genned PRs), will find it harder to become an SME.

If this policy is about trying to encourage human contributions and elevate human contributors over AI, it is not really achieving that and is potentially creating an impediment to that, simply through the existing project dynamics and what it will add to it as a consequence.

Reviewer 13: I'm personally against the whole: "For every non-AI SME we must add an AI SME" as that's just going to push out people who don't want to review or like AI.

Reviewer 5: > I'm personally against the whole: "For every non-AI SME we must add an AI SME" as that's just going to push out people who don't want to review or like AI.

In that case we need to come up with a clear solution, since otherwise if SMEs can unilaterally _block_ AI contributions on their areas of the codebase by refusing to review PRs the whole policy is moot, and we're wasting our time discussing it

Reviewer 13: > > I'm personally against the whole: "For every non-AI SME we must add an AI SME" as that's just going to push out people who don't want to review or like AI.
> 
> In that case we need to come up with a clear solution, since otherwise if SMEs can unilaterally _block_ AI contributions on their areas of the codebase by refusing to review PRs the whole policy is moot, and we're wasting our time discussing it

If an SME is unable to meaningfully say no, and is instead worked around with someone else being appointed just to say yes, then their no and refusal to review AI Generated Code is meaningless.

Reviewer 5: They should be able to say no (reject specific PRs) and delegate reviews if they don't want to do it, but it also wouldn't be reasonable to just say no every single time, instead of delegating to someone else who wants to review it. That would be following the word but not the spirit of this policy, i.e. malicious compliance

Reviewer 13: > They should be able to say no (reject specific PRs) and delegate reviews if they don't want to do it, but it also wouldn't be reasonable to just say no every single time, instead of delegating to someone else who wants to review it. That would be following the word but not the spirit of this policy, i.e. malicious compliance

No, that would be well within the spirit of this policy IMO. If Bevy is forced to accept AI contributions in an area despite the people who know that area well saying no, then saying no is meaningless and why even let SMEs say no? They will inevitably just be pushed out and replaced with someone who is willing to do so, and that will harm the community.

ETA: That would be basically telling SMEs either: "You must review AI, or find someone else to do so and leave." as there is a _strong_ bias towards those that actually do things, for better or worse. And as stated if one SME reviews AI stuff and another doesn't, the one that doesn't will be valued less.

Reviewer 5: > You must review AI, or find someone else to do so and leave

I don't see where the _and leave_ part would come from though. SME isn't a limited seat position, different areas have multiple SMEs, ECS and Rendering have 4 each, UI has 2.

Also it wouldn't even be the responsibility of the SME opposed to AI to personally find someone else to appoint as an SME or review the code.

Reviewer 13: > > You must review AI, or find someone else to do so and leave
> 
> I don't see where the _and leave_ part would come from though. SME isn't a limited seat position, different areas have multiple SMEs, ECS and Rendering have 4 each, UI has 2.
> 
> Also it wouldn't even be the responsibility of the SME opposed to AI to personally find someone else to appoint as an SME or review the code.

If they hold ethical objections towards using AI, and how much work someone does is the meteric, then if someone does more work as an SME by using AI and not refusing it, then they will be incentivized and preferred over the person who doesn't. This creates a necessarily hostile environment towards the person who does not use it, and either makes them have to change their stance to continue having their voice considered or avoid it and risk being sidelined. If they're then sidelined, you now get more pro-AI SME's who are incentivized to try and push for further allowing AI, which would make the situation with those who are against AI and already accepting a compromise with this policy to either try and stay -- while being sidelined -- or leave. Especially when there is the idea that: "If there an SME that won't. We must have an SME that will." Their consent here is functionally meaningless. 

It's like if someone's job offered AI as an option, but had performance reviews which advantages those using AI over those that don't. The ones that don't use AI can still work there, but they'll get worse performance reviews, and potentially fired eventually, while those that use AI (even badly) are allowed to stay and can eventually adjust policy so AI becomes mandatory. Even if the original intent was to provide a genuine choice.

Author: >  Contributors that do less aren't as valued as contributors that do more

This feels like it's a misconception of how open source, and Bevy work. We don't have a stack ranking system? I am grateful for whatever folks choose to do, and actively plan for folks to be missing or drop out for unexpected reasons at any time. This holds true all the way up to the maintainer level.

If people do good work, it will be recognized, regardless of their views on AI.

If we need more SMEs to ensure that AI or non-AI PRs become unblocked, we will appoint them. To the extent that we don't do this appropriately (and uh I sure think we need to do better here), that's a distinct problem.

Reviewer 11: I'd like to remind people in this thread that we don't have nearly as much power as you seem to think as SMEs. We can press the close button just as much as any other community members and we can't press the merge button. PRs don't _need_ a review from an SME to be merged unless it's a contentious PR. If an SME decides to never review any AI authored PR that won't really change anything. Community reviews will still be accepted and at the end of the day only the maintainers are allowed to merge code. 

This policy also isn't about blindly allowing LLM generated PRs. It's to encourage humans to work together as humans and to keep the quality of contributions as high as possible.

  Never serve as a direct pipe for an LLM: we want to talk to *you*, not to Claude-via-proxy!
4. **Large changes:** Please do not prepare or submit sweeping changes (by number of files touched, PRs opened or total lines of code) without prior discussion, even though AI makes doing so easier.
  Doing so will typically result in your PR being closed or ignored.
  These are a lot of work to review, attract merge conflicts quickly, and risk wasting resources for nothing.
  Just because you *could* generate a major feature in a single 10k LoC PR does not mean you *should* skip the design and consensus-building process!
5. **Refinement:** Do not dump raw generative AI output on other contributors: this is disrespectful of their time and attention.
  Use these tools to do things *better*, not just *faster*.
  Aggressively review and revise generated code before submission, working carefully to refine the design and polish the implementation.
  For written communication, seek to understand the output and synthesize it from scratch in your own words.

>   If you want to share exact output in a discussion, quote sparingly, with attribution.

Reviewer: Perhaps something also relevant to the code of conduct, but:

To me, "showing LLM output to another human without solicitation" as the `rust-lang/rust` LLM policy puts it (and forbids it in the general case), is rude and makes me disinclined to continue engaging in our community.

Sharing raw LLM output when solicited/when everyone consents to it is fine. But if allowing that is the intent behind this sub-rule, it is stated the wrong way around. If you want to share raw LLM output, you don't need your own consent, you need the consent of whoever you are sharing it with.

Reviewer 14: In context of written comms (discord, github): This sentence is worded in a way where it can be loosely interpreted as "allowed but try not to". I think the same rules should apply to sharing raw output in discussion as for writing documentation. Solicitation can be an exception but the wording should stress the common case to "do not share unsolicited, raw LLM output". Developing your own understanding, synthesizing it from scratch in your own words is the way to go.

weight = 1
+++

Bevy, as a project and a community, values:

1. **Community:** Collaboration and camaraderie between contributors
2. **Growth:** Developing understanding and wisdom within our contributors
3. **Quality:** Creating a reliable, high-quality foundation for others to build on

Our policy around AI use is designed pragmatically,
attempting to find ways of working together that will, in practice, honor these values.

In concrete terms, Bevy has the following norms around AI use that we expect our contributors and community members to uphold.
Some of these are firm boundaries, while others describe the standard of craft that we expect from our contributors:

1. **Ownership:** Contributors are wholly responsible for their own submissions, regardless of how they are created. 
  You must personally understand every line of code and documentation that you submit.
  Contributions must be original (or appropriately credited and license-compliant), well-made and crafted with care: something that you are proud of.

>   Do not include "Co-authored by Claude" etc commits; these will be caught by CI and automatically rejected.

Reviewer 15: I think it might be valuable for the whole document to separate out stewardship (and the theme of pride) from ownership. Maybe something like:

Stewardship: Put time and care into your work, and submit work you are proud of. You may not be involved with bevy forever, so be willing to contribute to our shared social and technical understanding of the project while you are here, and mindful of future maintainers.

I'm borrowing the phrase "shared social and technical understanding of the project" from the rust project's new AI policy. I find stewardship to be meaningfully different from ownership because one can only take ownership while they are involved with the project, and I have seen in the discord thread concerns about how AI code will effect the project after the person who submitted it has left.

Like with ownership, putting stewardship up front helps readers interpret where following norms are coming from.

Reviewer 15: maybe it would also make sense to include a line like "you should be able to teach people about your work"

Reviewer 15: The way I see the split between stewardship and ownership is that ownership is about responsibility, whereas stewardship is about pride and leaving a legacy

Reviewer 15: currently the ownership clause is handling both. 'something you are proud of' clashes with 'something you are liable for'

2. **Disclosure:** AI use, regardless of extent, must be disclosed, typically in the PR description, with context about how AI was used.
  This gives reviewers the information they need to evaluate both the output and your personal expertise effectively, and guide contributors to use these tools in a constructive way.
3. **Learning:** We want to help our contributors learn and grow.
  Work within the limits of your skills, and do not allow AI to substitute for understanding or thinking.
  Verify each finding yourself, carefully and critically evaluate plans and designs, and test things rigorously, both automatically and manually.
  Never serve as a direct pipe for an LLM: we want to talk to *you*, not to Claude-via-proxy!

> 4. **Large changes:** Please do not prepare or submit sweeping changes (by number of files touched, PRs opened or total lines of code) without prior discussion, even though AI makes doing so easier.

Reviewer 9: I think this prohibition should also apply regardless of whether the PR is produced by AI or not; I've seen plenty of unsolicited, hand-authored PRs which were too large to review.

  that were not AI generated?
- If there is a minimum threshold of human contribution to a combined work
  derived from AI-generated works for it to be considered copyrightable, where
  does that threshold lie, and is it consistently applicable to all types of
  contributions that Bevy Organization accepts?
- Does the local law in various countries and jurisdictions around the world
  provide consistent answers to all of the questions above?

Until there are well established answers to these questions, the use and/or
distribution of AI-generated code and assets may constitute copyright
infringement or may be subject to licensing terms incompatible with the FOSS
licenses used by the Bevy Organization.

Erring on the side of caution in light of a openly debated legal topic, all[^1]
forms of AI-generated contributions cannot be merged into repositories
maintained by the Bevy Organization. This includes both code and non-code game
assets (e.g. textures, audio, etc).

Any triage team member suspecting a pull request to be made primarily through
the use of large language models or other generative tools should mark the PR as

> `S-Nominated-to-Close` , upon which a maintainer can then review the PR for

Author: In practice, this was rarely done, even by members of the community with triage rights who were strongly anti-LLM.

> Bevy, as a project and a community, values:

Reviewer 16: One thing I'd like to suggest is taking some _good_ AI-assisted contributions to go along with the announcement in the next release post.  

Imo there's a lot of value to be had in:
* Debugging and reproducing issues during release candidates (pretty tedious work usually)
* Finding straightforward logic bugs in PRs
* Letting people explore outside their existing area of experience more confidently, without being as bottlenecked on conversations with existing contributors (it's helped me understand ECS internals recently)
* Letting already-experienced devs save time and contribute more to open source than they otherwise would be able to (In Solari, it's been very helpful for getting test scenes imported and setup from other engines, leaving more time for actual development)

I think it's worth highlighting the positives with specific examples.

Reviewer 5: Not sure if that would do us any good, ultimately.

_Accepting_ AI usage is different than _actively promoting_ it, and the new policy is probably already going to be contentious enough, I'd rather not have us be labelled as "AI-shills".

Edit: I _would_ be in favor of a retrospective analysis post, etc, for educative/instructional purposes, but not as part of the release notes itself.

Reviewer 16: That's very fair, I don't disagree.

> 1. **Community:** Collaboration and camaraderie between contributors

Author: Speaking in a more personal capacity, the question of what to do about AI has been tearing at me for months. I am not blind to the ethical objections around its use. Instead of boycotting it though, I've chosen to take another approach: using these tools for good.

While I, Alice, think that these tools are both effective (if used carefully) and morally worth harnessing while contributing to the commons, that stance is mine alone: not the project's, and not something that must be shared while contributing to the project.

For a long and heartfelt analysis of my personal feelings on the politics and ethical use of AI, please read [AI sucks. Hating it is not enough.](https://shaping.systems/blog/ai-sucks-hating-it-is-not-enough/). My personal political beliefs are provided here solely as context for my decision making, not as free-standing extra-textual rationale. I am one of the voices leading Bevy, but only one, and Bevy, as a project, foundation, and community, does not have anything resembling a coherent political perspective on this complex cluster of topics!

Finally: please be kind to each other, even as you disagree. I have made a lot of friends here over the years, and would like to minimize the number of them I lose.

> While we have a relatively permissive policy around AI use for pragmatic reasons, any use should be viewed through the lens of those values.

Author: **BROAD SENTIMENT THREAD**

Reviewer 17: I think this is a meaningful and positive step forward for Bevy and focusing on ownership, quality, communuication, people, relationships, and trust feel like the right things to be focused on.

Reviewer 18: Broadly speaking, I am against opening up bevy's AI policy to become more permissive of LLM usage (and especially diffusion model usage/any other forms of "art" generating models). If I had free choice, I would love to see bevy adopt a plain and simple no-AI policy such as Zig's, less so out of practical concerns and more out of political ones. Pragmatically, even a "never use AI" policy is about the same as "we don't care, we'll judge based on quality not process". The only real difference is, as mentioned, community values. 

That said, I don't hate this proposal. I have seen the general community stress that the old proposal was causing, and I think this proposal does a good job of trying to iron out pain points while keeping the overall technical quality and community relatively intact. 

Nonetheless, I already have been and am likely to continue withdrawing from public projects like Bevy for personal development. I do not want to build projects on a foundation that has been even partially built with AI, and so I won't, or at least I will do my best to minimize that within what is realistically possible.

Reviewer 13: While I wish bevy would take a stronger stance against AI, I'm at least able to live with this policy if enacted.

Reviewer 19: While I don't agree with a lot of your personal feelings on AI; I think it's hype is largely overblown and people are getting carried away with it, I do think this is a great AI policy for Bevy.

Reviewer 20: So far this policy seems well-situated to encourage humans to be good pilots. I have been wary of shame-based policies in the past and I hope this is more effective at helping the average dev become more capable rather than simply pushing them away (the juniors have to come from somewhere!)

Reviewer 21: I don't believe you will get anything valuable from this policy change. Sad to see another project buy into the hype.

Reviewer 22: I like this policy a lot and I think it focuses on the right things, especially that humans are responsible for the code they submit, disclosure about AI use and that AI writing is (pretty much) banned.

Reviewer 23: I think this is the strongest AI policy I have read yet.
I don't think this is applicable for all kind of projects and communities, but for Bevy I believe it will work.

Reviewer 24: I've been commenting on a fair few other threads, but I'll sum up my general views here. I haven't been incredibly involved in the development of this engine, but I love it and want the best for its future. For this reason, the ideal outcome in my mind would be to take a hardline stance against against any generative AI usage in the project, for multiple reasons:
* Generative AI is harmful and it is morally good for Bevy to oppose itself to it. This is the standard political/moral argument against its existence, you've heard this many times before and don't need me to reiterate it. The rest of these points will stand on their own, outside of this one.
* Slop begets more slop. Even if we only minorly change the tone of the policy, that will ultimately lead to more people who advocate for the usage of LLMs and similar programs to enter the community, leading to more permissive policies, in a long-run feedback loop. This is, in my view, a highly undesirable end.
* * Firstly, Bevy is a _game engine_, first and foremost. Games are a creative field, and LLMs oppose creativity. This is why the art community is so heavily against AI image generation. A game engine that is supportive of generative AI is a game engine that is opposed to its core purpose.
* * Secondly, as mentioned, artists are by and large opposed to genAI. You cannot support a games industry without artists, and so if Bevy adopts a stance that accepts genAI, it prevents itself from becoming a truly relevant game engine.
* * * One counterargument to this point would be to indicate towards the success of engines such as Unreal, with their integrated generative AI tooling. This is because Unreal is a game engine targeted not at artists or developers, but at game studios. Corporations by and large love anything that has the label of "AI" attached to it. Thus, adding more "AI" features is a good move for engines targeting a _corporate_ industry. Bevy does not target a corporate industry, and never will until it becomes mainstream in the indie space, and it cannot get into the indie space with an accepting stance, meaningfully.
* * Of course, I cannot predict the future, it is not guarenteed this will ever happen. Nonetheless, it is a real concern, and one that can only be addressed at its root: making it clear that generative AI contributions are not welcome in the project
* And one note, to address the concerns of unenforcability and distrust. These are issues that still exist no matter what, unless you openly accept generative AI contributions entirely. You can never truly know if a person used LLMs despite saying they didn't, the only "solution" is to simply not care. Writing a lenient policy does nothing to change that. As for distrust, while making a policy's stance lenient does not fix it, simply changing attitude and not being overly litigious trivially fixes this. Assume good faith, that someone is not violating the rules, unless they show they are. We will never be able to catch all AI-generated commits, and that should not be the goal of a hardline proposal. It should simply be to make clear that they are not _welcome_, so that any good-faith individual who would otherwise use generative AI either does not do so in their contributions to Bevy, or leaves entirely.

As I have said before, I love Bevy and all it has done. I desperately hope my fears are unfounded, but in my personal experience with communities who have gone a similar route, it is not, and I would hate to see that happen to this engine. Thank you for hearing me out.

—Hannah

Reviewer 25: The values listed form a good basis for the policy, which I think is very well written. Also the new appendix section is a good addition.

Reviewer 26: I'm going to reiterate and expand on some of what I said on discord here:

- We shouldn't underestimate just how widely disliked AI is in some parts of gamedev. I'd seriously estimate that an overly permissive AI policy (not necessarily this policy, but maybe) could cost >50% or even >90% of potential users. People can just write projects off entirely over things like this.
- Part of how this plays out concretely is, if I want to recommend Bevy to people, I would now either have to:
  - disclose this policy to them, which I feel will likely kill interest immediately,
  - not disclose this policy, which feels dishonest as I expect it to materially affect their decision,
  - not recommend it at all.
- I can't think of anyone I know in gamedev outside the existing Bevy community who I would expect to react positively to the policy, which means either I or the Bevy community is in a bubble, and I don't know which.
- It's very difficult to win back user trust after it's lost, like what happened with Unity and their pricing model changes.
- A policy of requiring disclosure means inevitably endorsing AI written code, which would also be nearly impossible to reverse later.
- It is impossible to enforce a ban on AI code, and I don't think it's even worth trying to, apart from banning co-contributed commits and the like. However, Bevy's policy could simply both discourage AI and also refuse to police it.
- I'm not an AI opponent, I use it myself in other contexts, and I'm a big fan of Bevy as an engine. I'm going to keep using Bevy whatever the policy is, but emotionally this feels like watching someone carefully preparing to shoot themselves in the foot.

Reviewer 24: Strongly agree with those points—a policy that is simultaneously against the usage of AI in Bevy contributions and one that acknowledges this is something that cannot and should not be reasonably policed, and that good faith should be assumed on the part of contributors, is a policy that addresses every apparent issue with the previous policy, while preserving the integrity of the Bevy project itself.

Reviewer 9: I want to give my perspective as someone who gets a lot of PRs assigned to them to review.

A few weeks back, I was asked to review a lengthy PR. I checked it over very carefully - it was actually quite good, well structured, well documented. Only - it was a little *too* well documented. I didn't see any of the obvious AI "tells" (no use of the term "load-bearing"), nor did it suffer from tendentious over-explaining ("LLMsplaining"), but the comments were better written and lengthier than I expect from most of our contributors (as someone who writes loquaciously, I am sometimes guilty of this myself). There were also a few other subtle signs of LLM-generation, but nothing that would detract from the overall quality of the PR.

So now I have a dilemma: I strongly suspect the PR was LLM-generated. Do I reject it out of hand, without proof? Do I ask the author if it was generated? (with the implicit threat that if they say "yes", the PR will be closed, and if they say "no" now I have to wonder if they are lying or not, so we're screwed either way). Should I just say "not my job to police AI usage" and let it go through?

My point is this: while the arguments about long-term consequences (whether the contributor will stick around, preservation of tribal knowledge, community cohesion, etc.) are important, ultimately the policy has to boil down to a set of actionable decisions that reviewers can take in the moment.

And yes, I know the policy has to be more aspirational than actionable, but we still have to be able to take actions.

Reviewer 5: > a policy that is simultaneously against the usage of AI in Bevy contributions and one that acknowledges this is something that cannot and should not be reasonably policed, and that good faith should be assumed on the part of contributors, is a policy that addresses every apparent issue with the previous policy, while preserving the integrity of the Bevy project itself.

This is essentially a "don't ask, don't tell" policy, and I believe it would be terrible for the project's long term morale and community building

Reviewer 2: (General Sentiment) Not a bevy contributor myself, I am just someone that has closely followed the project for years, and has been waiting for bevy to mature to a stage where I can use it in my projects.

The reason I lurk and care about bevy at all is because what bevy is building is the highest quality game engine, with great decisions made around its design and code quality. Anything that undermines that quality bar is for me is my biggest worry. I do not believe Bevy's quality bar improves with this change, and I do not mind a slower "velocity of features" by banning AI. I am also a realist, my head is not in the sand, I understand that AI works and can write good code. But the issue is far more subtle. It's the systematic application of it over time that slowly eats away at a codebase, ship-of-Theseus style, and undermines the quality in a slow and hard to explain way. Anyone that has worked somewhere that has been primarily vibe coding changes for greater than 12 months ought to be able to see what I mean.

Universally, no matter how much I use AI in my projects or at work, it does things I wouldn't have and I am dissatisfied with its approach. I would be comfortable with bevy allowing AI for the following cases:
* minor bugfixes 
* initial ideation of code to vibe check a design
* largely mechanical refactors (but why then can't you use your IDE and/or vim motions + macros?)

However for non-ideation feature work, I find as a general rule that AI tends to:
1. undermine the quality of the code (not in functionality or bugs but in structure and taste)
2. require just as much time to shape from its initial output into what I want as if I had coded by hand
3. Incentivizes myself and others to take kludges,  lower quality, and operate outside our own capacity
4. decreases my understanding and ownership of my own code
5. removes my ability to lean on others (coworkers, community members) for their expertise and understanding, since it is now unclear to what level of care or effort they put into the output. It breeds distrust.
6. it produces a general sense of alienation from the craft, much in the same way it alienates an artist from art.

The new policy tries to mitigate these risks by saying essentially "be careful". But telling people to be careful isn't the bar bevy should have: bevy should actively discourage AI generated submissions, because the incentives to degrade quality over time are just too high, the harms to our craft, art, ecology, and labor are massive, and the benefits to productivity are not large enough to justify those costs. (Not to mention, the new policy is [similarly unenforceable](https://github.com/bevyengine/bevy-website/pull/2551#discussion_r3723269703)) 

Finally, I want to echo what @tmstorey and @PROMETHIA-27 said. The alienation of artists is real and *will* hurt Bevy's adoption. Not to mention, what about the alienation of us programmers that consider code to be a craft and not a purely technical endeavor? As a personal anecdote, it is *exhausting* having to deal with massive quantities of AI generated code at work, and a lack of appetite for trad coding due to time/labor cost. FOSS is supposed to be a respite from the proprietary world, why are we seeking to copy its playbook?

I sincerely hope this policy is not adopted.
-Ryan (TheButlah).

Reviewer 2: <details>
<summary>As an aside:</summary>
I think there is a frustrating asymmetry here in having the original PR comment able to be presented in one comprehensive chunk with the benefit of prior discussions but responses to it must be separated and chunked into threads. It makes it extremely hard to formulate a coherent argument against the new policy. However, I'll do my best to abide by the moderation rules as presented, and link to threads on specific grievances from this post.
</details>

> While we have a relatively permissive policy around AI use for pragmatic reasons, any use should be viewed through the lens of those values.

Author: Some folks have suggested that we should include more of the history / motivations around culture / enforceability into this document itself, rather than leaving it entirely to the PR description. Doing so would help contextualize Bevy's choices and values, and why we've selected the policy we have.

I am somewhat loathe to do so, because it's not actually useful to future contributors, and I would like to keep this policy focused on "things I actually immediately need to know to contribute to Bevy".

Reviewer 5: Agreed on this, but maybe let's add it to a (collapsible?) section at the end.

Reviewer 13: I do think the history is actually extremely important here, because this policy only makes full sense when paired with it IMHO. "Why don't you just ban AI" will likely come up again in the future as well, having a section over why Bevy adopted this path is probably the best. 

Besides the existing Appendix already does get away from that a little bit.

Author: How do y'all feel about an Appendix on History / motivation then? Please use 👍🏽 and 👎🏽 reactions on this comment to give me a sense on this narrow question; I promise it won't hurt my feelings.

Author: I've added an Appendix on History and Rationale now!

Reviewer 25: The history/rationale section as it reads now is a very good addition I think.

> Accordingly, Bevy has the following norms around AI use.

Author: Many of these items are no more enforceable than the existing ban on AI-generated content! Does that make them bad rules? Is there a *point* to having rules at all unless they can be automatically and impartially enforced?

I've spent a lot of time thinking about these questions, and my answer is that these debates are fundamentally about norms and values, not rules in a legalistic sense.

Bevy, as a community, should pick its norms and values deliberately and explicitly.
Those listed in this policy are the values that I hold for Bevy, and the norms I want to see.
Hopefully you agree, or can at least live with them and understand why we might choose those <3

Other communities should do likewise, regardless of the values you hold. 
In retrospect, I think that some of the problems with the Endless AI Discourse was *because* the initial LLM policy was fact and rule-based, rather than values and norms. We wanted to land something clear, that we could immediately point to in order to deter slop, and do so quickly, cutting anything that folks might disagree with in an effort to find consensus.
IMO, doing so simply put off the hard conversations, and left everyone feeling like the *actual* questions about "what do we want Bevy to be" were completely unsetttled, and thus a battleground which must be fought over.

Reviewer 2: Re: concerns about enforceability. I agree, this policy isn't enforceable either. How do you demonstrate someone had "ownership" over their code? How do you know that the ai generated PR couldn't have been done better or differently if the driver had been more closely at the wheel? Any AI policy is inevitably unenforceable as the model outputs are generally undetectable. I understand that we want a concrete "this must be the absolute metric to weigh a moderation decision on" but it's not possible.

> I think that some of the problems with the Endless AI Discourse was because the initial LLM policy was fact and rule-based, rather than values and norms. 

I also agree on this point. But the new policy throws the baby out with the bathwater: due to the *inherent* lack of enforceability in this space, it pivots too hard and says "well we officially are ok with you using AI just use it well". That's not something bevy should tolerate or endorse (more on that in my general sentiment post). 

The most important thing that a policy does is it communicates the communally held standards of behaviour we hold each other accountable to. And instead of establishing the lack of tolerance for AI, the new policy completely gives up and says "use AI just use it good". 

Regardless of how you feel on AI, it should be clear that this is a massive pivot and that enforceability is not a justifiable reason to change the actual standards of community behaviour, it can only change the way those standards are policed and enforced.

My preferred change is one that weakens the enforcement to assume good faith and not micro police, but not to change the actual community standards of behaviour

Author: > My preferred change is one that weakens the enforcement to assume good faith and not micro police

This is in effect what our policy enforcement collapsed to. It was very stressful, as described in the PR description and history section.

>   This gives reviewers the information they need to evaluate both the output and your personal expertise effectively.

Author: The inclusion of the evaluation of personal expertise here is unusual but deliberate. As a community, we need to be able to assess who has the judgement and expertise required to lead Bevy in the future, and an understanding of how much to implicitly trust design choices and opinions from each contributor.

There's no bar for skill or expertise (everyone starts somewhere), but preserving the signal here is very important to us!

Reviewer 20: I feel like my personal bar here is "can you, as a contributor, discuss your thought processes and rationalle normally with other humans?". Unfortunately, this kinda breaks down because generated content is too similar to crafted content when groomed properly; however, maybe this in itself is a decent bar, as to properly groom a diff you generally need to have a good understanding of what it does and why.

Review is hard lol

Reviewer 5: If it _must_ be disclosed, what are the consequences for failing to do so?

Reviewer 13: > If it _must_ be disclosed, what are the consequences for failing to do so?

I would assume failing to do so would have consequences ranging from a warning to being barred from the project depending on circumstances.

Author: Yep: that sort of flexible-and-serious but not draconian approach to moderation is what I would like to see here. See the point on Good Faith. We should begin with that assumption, and create incentive landscapes where people are encouraged to do the right thing, but that does not require us to be endlessly naive.

>   Trivial use is defined here as use as an analog for an existing tool such as a search engine, grep or interactive autocomplete.

Author: Is this a sufficiently clear definition? Is this the right place to draw the line? My goal here is "all AI use that would be helpful for me to know as a reviewer is disclosed but no more to reduce noise".

Reviewer 5: Where would something like a preliminary code review by an LLM fall in?

Author: I would say "this is useful to disclose, with context about which issues you have verified or are uncertain about yourself". It gives us information about origin and contributor expertise, and helps people form opinions about which practices work and do not work well, from an engineering perspective.

Reviewer 27: I would suggest simplifying this to "any AI use must be disclosed". Disclosure is easy, so it doesn't really add a burden to authors, and it removes all ambiguity from the clause. For complete transparency, I have also gotten into the habit of disclosing when a PR is hand-written in projects where disclosure is encouraged.

Author: Hmm. That would be simpler. Also maybe a bit noisier. I don't feel super strongly about this? I think you're probably right, but I want other folks to chime in and to sleep on it before I make the change.

Reviewer 5: Maybe we could have like a multiple choice thing in the PR template, like:

> ### AI Disclosure
> Please select all that apply:
> - [ ] Research
> - [ ] Autocomplete
> - [ ] Agentic/Autonomous Code Generation
> - [ ] Code Review
> - [ ] Mechanical Edits / Automation
> - [ ] Proofreading
> - [ ] Translation

Would even be cool to get some stats down the line

Author: Mhmm, better templates for PRs and issues seem like a low-cost intervention. A review checklist is also on my long-term wishlist, and I think would be really useful.

>   "Co-authored by Claude" commit messages are not sufficient (or particularly helpful): lay this out in your own words in the PR description.

Reviewer 28: One thing that is not mentioned is whether AI-authored commits are allowed (and thus making e.g. Claude a co-author on PRs).

Discussed on Discord [here](https://discord.com/channels/691052431525675048/1533897174587277424/1533904923698401420), relevant comments copied here for non-Discorders:

> Jondolf
> is AI co-authorship in commits allowed? don't think I saw that one mentioned
> if not, might want to mention it and maybe add a CI check, like the one I wrote ([link](https://github.com/Jondolf/ai-commit-check))
>
> Alice
> Fine but not sufficient or particularly useful. I chewed on your CI check for Avian a bunch and ultimately decided that while I think the co-authored by statements are annoying ads, they're not worth burning CI minutes / contributor frustration on
>
> Jondolf
> Personally I don't really like seeing a lot of Claude commits in projects / it makes me somewhat lose interest, and so I prefer just not allowing them. It also filters out the sloppiest slop (if you're letting an LLM make actual commits for you without a human first going through the changes, that's in general too far for me personally) and gives a clear signal that those kinds of contributions aren't wanted, even if the policy is generally open to some AI assistance. But yeah, I can see allowing them being a reasonable stance too with the more lenient policy, one could argue that disallowing it just encourages hiding AI usage
>
> Alice
> Mhmm. Similar concerns have been raised about a AGENTS.md file: signalling quality vs. encouraging compliance with the policy

Opening this thread for discussion as per Alice's request :)

(Edit: lots of other discord comments too after the ones above, just copied the start of it)

Author: My feelings here are very weak, for the record: if y'all think this is worth including I can make a change to the policy to make that a rule.

Reviewer 29: I think we should disallow the co-author lines in commits. I think it could help prevent automated PR spam. And as cart mentioned on discord it could be sending the wrong message.

But as others have pointed out, it's probably not enforceable to ban agent-authored commits _entirely_ since you can just remove/disable those lines. So maybe just banning the commit line should be enough

Reviewer 13: I'm rather supportive of banned AI authored commits at the very least, as it discourages AI spam as mentioned before

Author: My primary motivation for restricting this is that I fundamentally view it as impolite advertising. Disclosure is good, but context-free company-driven defaults are a bad way to do this. I would also be annoyed if Apple introduced a "Sent from my iMac" line as part of commits!

Reviewer 5: `Co-Authored-By` sucks, and is indeed like an ad.

I like what Linux is doing (`Assisted-by`)

https://github.com/torvalds/linux/blob/848acc8ffe1b7cd5f1bf427b93069becfebc2c9d/Documentation/process/coding-assistants.rst#L56

Author: I'm adding a commit to forbid "Co-authored by:" :) Thanks for the discussion y'all.

I don't like Linux's Assisted-by approach for Bevy because it emphasizes commit messages in a way that doesn't line up with Bevy's review / PR workflow very well, and asking beginners to do anything even remotely complicated with git is a recipe for distaster :p We've long accepted "eh, just squash and merge" as the pragmatic standard, even if on some level structured clean commits are *nicer*.

> 3. **Learning:** We want to help our contributors learn and grow.

Reviewer 16: (Not seeing a thread on Learning yet)

I'll be the first to admit that this is very tough. AIs let us branch out easier and get more done, but at the cost of understanding. Sometimes I find this tradeoff is fine - I have limited time, and sometimes I just want to get something done.

But even those small things are what contributes to growth, and over time it's very easy to realize you've achieved a lot, but understood very little.

I don't have a solution, nor am I good at this myself, but I encourage people to try build understanding for the long run, even at the cost of short term tasks.

If you plan on being a long term contributor (and very few people in my experience tend to plan for that to happen, rather than simply finding it fun and keeping on contributing), I encourage you to keep in mind that OSS is more of a marathon than a sprint.

The people I find most valuable in conversations about PRs are the ones with the most context about how and why things were built the way that they were in Bevy, and how that informs future plans.

Reviewer 20: I have personally found a lot of value in using AI to create proofs of concept, then taking that and, with the gift of hindsight, reimplementing it manually. Gets those rote memorization muscles going. I know this likely isn't an entirely common approach, but maybe one Bevy can encourage.

Reviewer 5: This is also an area where AGENTS.md could be really useful, for example, look what Stanford is experimenting with for their CS courses:

https://github.com/stanford-cs336/assignment1-basics/blob/main/AGENTS.md

Reviewer 11: Having an `AGENTS.md` file makes the repo look like it's very AI friendly which isn't really the stance of this policy. I'm pretty sure many people that would agree with this policy would still avoid the repo before even reaching the policy if they saw that kind of file.

Author: Mhmm; I'm caught between "I think that this would send the wrong message to potential contributors" and "I think this would help contributors follow the policy / make better contributions". On the balance, I would prefer to avoid adding anything like AGENTS.md, especially in the short-term, in favor of improving human-facing documentation and cross-linking.

Reviewer 30: I'm very skeeved by adding an AGENTS.md. Even when projects use jokingly as an anti-ai message, I immediately take note and just assume it was vibecoded. That assumption might change as LLMs change, though.

Reviewer 24: I agree with the proposed opinions on AGENTS.md and similar files: Seeing a file like that almost immediately makes me want to ignore the repository, and we do not really lose anything by keeping all LLM policy in an appropriate policy file here. Additionally, the same is the case in the opposite direction, to people who use LLMs to generate slop, having an AGENTS.md makes the repository seem like it approves of that. Omitting one altogether could deter the worst offenders.

Reviewer 5: Hmm I think it's possible to have a cargo build script that detects various coding agents via env variables, install paths, etc, and creates a (.gitignored) `AGENTS.md` file locally. This could allow us to only have the file in context for those that use coding agents, while not having it prominently displayed at the root level of the repo perhaps?

I do think the signal will likely lose its meaning as more anti-AI projects add AGENTS.md explicitly disallowing agents, (kinda like robots.txt) and various non-slop projects add it, so this is largely a temporary situation.

> 4. **Large changes:** Please do not prepare or submit high-volume changes without prior discussion, even though AI makes doing so easier.

Author: This has been true (and a problem!) for years, but is included here because AI makes the potential problem much worse!

Reviewer 16: To add to this, even without AI, we have more code than we do reviewers !

It's unfortunately true that the best way ime to get PR reviews is to ask people you know have experience in the area on discord, and get prior buy-in. In an ideal world we would be able to triage and review every single PR that comes in each week, but we don't have the time or people for that.

I wish we had a solution to this (and I've been a proponent of auto-closing inactive PRs in the past, which I know is controversial).

> 4. **Large changes:** Please do not prepare or submit high-volume changes without prior discussion, even though AI makes doing so easier.

Reviewer 31: I think "high-volume" is unclear here. It makes me think first of "high volume of PRs", i.e. sending many PRs. How about "PRs changing large amounts of code"?

Author: ```suggestion
4. **Large changes:** Please do not prepare or submit sweeping changes (by number of files touched, PRs opened or total lines of code) without prior discussion, even though AI makes doing so easier.
```

> 6. **Writing:** Everyone is sick of reading LLM-flavored text, no matter their feelings on its use.

Author: Note: on a technical level, capabilities and defaults here may change! There's nothing in the technology that says LLMs *have* to talk like an insufferably smug customer service philosopher!

Nevertheless, I think that this is worth including even if that changes, largely because of the concerns around "writing is thinking", outlined in the next bullet point.

Reviewer 17: > 6. **Writing:** Everyone is sick of reading LLM-flavored text, no matter their feelings on its use.

I both personally agree with this, and I think it should be removed.

I do assume that LLMs will get better at writing such that they aren't as long winded, repeditive, and annoying sounding, at which point I expect them to be even more useful tools for drafting and writing.

But the reason I think it should be removed is that I worry it leaves open some of the negative behaviors that we've seen.  "Used an emdash?  Must be an LLM!"

Reviewer 13: I'm against dropping this for the simple reason as follows:

This sends a signal about the kind of community and project Bevy is. It may not be strictly enforceable, but just having this here sends a signal about what Bevy -- as a community and project -- endorse. I also think witchhunting will be less prevalent with this new policy as it addresses some of the more fundamental reason the witchunting happened in the first place.

Reviewer 21: > I do assume that LLMs will get better at writing

Is there any data to support this assumption though? It's an autocomplete, if it was going to improve with training it would have improved a few exabytes of data ago.

Those behaviors won't disappear if this point is removed, and having it at least sets some expectations.

Reviewer 20: LLMs so far are abysmal at creating human-lexable technical writing in any capacity, but so are noobs. I think this could probably safely be omitted from this policy in favor of a general "we should write technical documentation/comments that don't make people's brains melt"

Reviewer 13: > LLMs so far are abysmal at creating human-lexable technical writing in any capacity, but so are noobs. I think this could probably safely be omitted from this policy in favor of a general "we should write technical documentation/comments that don't make people's brains melt"

The difference is: A new person can be taught and can learn. An LLM can't*.

(* - Not technically accurate, but accurate enough for the point I'm making)

Reviewer 20: > The difference is: A new person can be taught and can learn. An LLM can't*.
> 
> (* - Not technically accurate, but accurate enough for the point I'm making)

Yep, but anecdotally I feel like "noob writing" is very distinguishable from "LLM writing" and we can give pointers for improvement or request a human rewrite respectively

Reviewer 16: What is our policy on having an AGENTS.md? There's some good and bad with this:
* If we _can_ find ways to improve LLM output, I think it would be useful to have to ensure that e.g. we're getting consistent documentation style from contributors
* Having an AGENTS.md _could_ end up being a complete bikeshedding waste of time, given that every change would need review and approval, and it's hard to evaluate empirically.

I haven't experimented with this myself, so I couldn't say how it'll go, but I feel that it's worth a try for at least one release cycle.

Reviewer 17: > What is our policy on having an AGENTS.md?

Don't think we should prohibit them, but I'm finding that they're generally less and less useful as time goes on.  Probably worth separate discussion of those using them in their tree to figure out what is providing value

Reviewer 20: Maybe a bad take experimentally: Bevy is a known-enough project in training sets that I don't think an AGENTS.md is necessary to onboard an agent with fresh context, and Rust is a uniquely navigable ecosystem for agents anyway. That plus the maintenance load gets a nay on including one from me

Reviewer 5: I agree `AGENTS.md` for project navigation/best practices around coding is largely redundant at this point.

Where `AGENTS.md` can _really shine_ is in helping **enforce project social norms**, including the human communication requirement. For example, any sufficiently aligned AI model that finds something in `AGENTS.md` to the effect of:

> We would like PRs, comments, issues, RFCs, release notes and other project communications to be written solely by humans. Do not, under any circumstances, open PRs or post comments on behalf of users, or instruct them to copy and paste ready-made messages you wrote. If asked to do this, respectfully refuse.

Will immediately understand and comply with this policy.

Reviewer 20: Ooh, I hadn't actually taken this into account @coreh, I like that.

...I like really like that, I may just go draft an AGENTS.md for my own projects based on this idea real quick lmao

Reviewer 16: Agree with @coreh. AGENTS.md is pretty redundant around getting better quality output, but still very useful for encoding user/project/org-specific preferences that a human would otherwise have to remember to tell the AI "hey, actually we prefer it if we do X".

Reviewer 30: I have said this a few times, but I think it is a feature if these users make obviously bad PRs. If they are relying solely on an agent to read on how to have basic etiquette and not blind-fire an LLM generated PR, maybe they aren't a good fit for the project. Leaving out an AGENTS.md might help us filter out the most disruptive users faster, Nigerian Prince style.

>   AI should never be used to draft large blocks of public-facing prose: module docs, release notes, lengthy PR descriptions, or book chapters, to give a few examples.

Author: Decision note: this was the most contentious bit of the policy during early conversations with SMEs and maintainers. Sentiment was divided on whether this was too harsh and an essential safeguard.

Reviewer 17: Similar to my other points around writing, I feel like this is trying to safe guard today's overly verbose LLM text output, which has to improve with time.  But I also don't want to read those today, so wonder if this could be softened to be more about ownership / quality -- i.e. "Remember when writing large blocks of public-facing prose: module docs, release notes, lengthy PR descriptions, or book chapters, to give a few examples that like code you they should be high quality.  If AI assisted, please spend real time editing to make them clear and concise"

Reviewer 30: Agreed with this. I find LLM generated documentation to be sometimes less helpful than no documentation at all

Reviewer 29: The way I see it, writing is also a form of art that I believe deserves to be protected from LLM generation. However, I also see arguments against that opinion (e.g. technical writing differs from creative writing, etc.). So if we were to allow it, I agree that it should definitely come with the caveat of requiring full understanding and being human-edited.

Reviewer 30: More pragmatically speaking, programmers are likely far better at writing and recognizing good code as opposed to good writing and documentation. While LLM output can be helpful if monitored and edited, I am skeptical that the average PR with LLM generated docs would meet a baseline bar for quality.

Reviewer 18: Personally, I think this is a "bare minimum" restriction. For bevy to start replacing its (very high quality) public-facing writing with LLM-drafted text would be tragic, and severely decrease my opinion of the project, and I would imagine many others agree.

Reviewer 13: I personally think this is fine. LLMs are merely statistical models, they can do work, but writing is explicitly for communication between humans. Outsourcing that work to LLMs is fundamentally anti-humanity and has an unacceptably high chance to degrade the community around bevy IMHO.

Furthermore, as noted Bevy wants to appeal to artists and this is likely already to at least cause some issues with artists, allowing AI in writing text (especially publicly facing text) would likely drive away many, many more.

Reviewer 16: Even for non lengthy PR descriptions, e.g. just filling out the template, I would rather a human written description. 

While an AI could write a good description, if prompted to (same issue as the larger discussion around docs/etc), I feel like the best quality for PR descriptions is just a human writing it by hand, even if it's kinda badly written. It's almost that the less info in the PR description, the better, and I want to encourage people _not_ to dump AI summaries, even when condensed, as a summary is a bad use of a PR description.

As an aside, perhaps we should update our PR template, especially if we do forsee AI being used to answer the template questions.

Reviewer 32: I've often found that when writing anything really (especially pages for the Bevy Book), editing takes just as long (if not longer) than writing itself. The time it takes to edit can also dramatically increase once you're dealing with something you didn't write yourself. I don't see how LLM generated text doesn't fall into the worst of this pattern.

With this, keeping all human-facing writing that Bevy puts out human-written is the most sensible path in my opinion. Use AI for basic corrections, grammar checking, it might even give some decent feedback about verbosity or clarity, but letting AI write for us will not benefit us in the long run.

Reviewer 20: A thought from some discussions we've had in the JJ server:

> You can't easily take a thought you've been bouncing back and forth with an LLM for hours and export it raw for other humans to ingest. It needs to be re-encoded in human speak, by a human, with human brain sized scope to be properly received and understood

Reviewer 16: RE: Documentation

I actually _would_ like to try experimenting with AI for docs. We have a lot of good contributors, but very little time or energy to maintain good docs. 

As someone experienced with Bevy's rendering, but simply doesn't have the time or energy to write docs I feel are high quality enough to contribute (and I've tried before! https://github.com/bevyengine/bevy-website/pull/2195), I'd love to be able to use AI to improve our docs.

AI could really help get over the barrier of "just get anything on paper to start", ensure consistent style between different contributors, and make sure we cover the different forms of documentation https://diataxis.fr in a way that we frankly don't have the human budget for.

The risk is of course that AIs can generate some _terribly_ wordy and weirdly written text without enough human authorship, and we'd still have the same problem we have now with existing doc PRs and past attempts: They are large and take quite a lot of time and discussion to get merged, relative to their value. Writing docs faster won't be much helper if we still need to get them approved at the same speed.

Reviewer 20: @JMS55 In the spirit of the proposed policy I would absolutely be fine with a "here Claude can you spray&pray a doc outline for me to fill out", it's a lot more "what concepts need to be covered" and not "please write documentation for other humans for me". Plus you can have it cross check your written docs against the code and help point out gaps that may trip up readers

Reviewer 16: It might depend on the style of documentation you're writing? For longer book-level docs, yes absolutely, I would not want an LLM generating every word.

For simple function-level docs, I feel like we might get more value out of allowing it and just having a human giving it a once-over.

I am undecided, partially because a lot of doc work is boring drudge work that otherwise wouldn't get done.

Reviewer 20: I will say I have hand-rewritten nearly every function doc that Claude has ever written me, but at the very least it was a starting point that forced me to say "ew no that's so incorrect, let me make that better" a la [Cunningham's Law](https://lawsofsoftwareengineering.com/laws/cunninghams-law/)

Reviewer 33: I think all the proposed policies around written content (including those below) are reasonable. Human-to-human communication is fundamental to maintaining human communities and fostering human growth; allowing  LLM-*generated* or *authored* writing undermines those core values, in my opinion. Being able to articulate your ideas / code / design in your own words is a sign of ownership, exertion, and understanding; it's a test that you're (A) thinking and (B) learning, not just passively ingesting.  It's all too easy to convince yourself that you understand information you've casually absorbed, but if you can actively, coherently express the concepts, it's a very reliable positive learning signal. Writing is a learning pathway and I think it's very important for a community to preserve all of those pathways, since they're probably the mechanisms by which new contributors improve and a project remains healthy long-term. A restrictive writing policy also encourages good-faith contributors to develop these skills, which has social value in and beyond Bevy.

Reviewer 32: @JMS55 Normally I'd also be of the opinion to just get something out there to start, but I think we also have to contend with the fact that unless it's something kept well out of the public's eyes (i.e. entirely off of the site/outside the docs/even avoiding draft PRs), we run the risk of users finding it and using it. The best case scenario is that they get the gist of it and can do their own research on the areas that the AI might've missed, but even then they might get confused if they return later and see the whole page rewritten/heavily edited to now have differently worded or entirely different information. Worst case scenario is that the AI misses or incorrectly states something (or all the things) relevant to actually understand the topic. I see the benefit from a maintainer's point of view, but I don't think sacrificing the UX is worth that trade off.

Reviewer 9: I want to repeat a point that I made privately: there are some cases where LLM-generated text is *better* than what a typical human could write. Specifically, tabular data or highly repetitious content is the kind of things that humans are typically bad at - after all, part of the reasons why computers were invented in the first place is to deal with highly repetitive tasks.

This doesn't mean that I want to read LLM-generated narrative prose. But for something like a TODO list, I am fine with it being machine-generated. For one thing, such lists almost always have items in them that I probably would have missed. Of course, I am going to review the list - and possibly add some items of my own. The question is, whether I'm violating the policy by pasting that list into a github ticket.

Similarly, when I'm using the LLM as a "smart regex" - e.g. "find all uses of this deprecated symbol and group them by migration difficulty".

Another example: if I ask the LLM to generate an outline, and then fill in all the paragraphs myself, is that a violation? I mean, I write a lot of text, but even I sometimes suffer from "blank piece of paper syndrome".

Reviewer 30: Another anecdotal point worth mentioning. I very recently cleaned up documentation that was LLM generated (or at least assisted). It ended up taking far longer than I expected and was tedious even for a short amount of text to the point where it likely would have been faster to start from scratch. 

I first had to understand what the LLM was trying to say, and then translate into something human readable (as output is terse and often produces actual nonsense that we tend to gloss over and read as "verbose" -- the perspective is **way** different casually reading vs. editing. It's hard to see how many problems LLMs generate just by passively reading/reviewing.) 

---

I'm hesitant to ever be OK with allowing the direct output of LLMs in prose, even for TODO lists, because it's easy to settle for the given output. I'd much prefer reaping that benefit by using LLMs to research without copying their direct output. Often I find that the added friction forces me to think a little  harder about if the text I'm copying fits grammatically, etc.

I'm OK with formatting and table creation, as long as the content is human authored. Generally, if the action an LLM does can already be done in another non-AI tool, I don't have a huge issue. It's likely undetectable anyways.

Reviewer 13: I'm personally more persuaded by @piedoom's argument here. Although if something a can be done by a non-AI tool, I don't know why you wouldn't just use the non-AI tool in the first place tbh.

Reviewer 27: Interesting, this is the part I agree with mostly strongly. I feel like AI-generated prose if often considerably worse than none at all. My main issue is that AI-generated text is sometimes subtly wrong, conveys intent that is not actually there, and is very hard to verify for correctness (unlike code). That, and it's needlessly verbose.

If it is heavily edited after generation then it's probably ok. But in that case it's also likely indistinguishable from human-written prose.

Reviewer 12: Given my own experiences seeing what AI gen'd documentation looks like, and the awful time I've had trying to follow it and being unable to get things done because it kept getting details wrong (and this is also based on usage in the company I work at), I am very, very against AI usage for documentation.

Even if it could generate text that wasn't awful to read, there is a lot of nuance in technical documentation, knowledge & *experience* that an LLM would just not have access to by virtue of being a model. That requires a human to write these things down, and bad human docs is usually better than bad LLM docs, because it is easier to clock it as bad & it would still have *some* actual nuance and experiential notes than LLM text would just not have.

This is one area that absolutely *needs* human collaboration, not AI assistance.

Reviewer 5: Also if you wanna read AI-generated documentation you can just ask for your AI model of choice to write that documentation on the fly for you, so not much is lost.

I suspect a UI for LLM generating docs lazily/on the fly as you navigate through them either already exists or will show up soon.

Reviewer 23: I personally rather want an AI generated external documentation than no documentation at all.
I think there's a bias that every person reviewing here knows enough of the project that every subpar documentation is justifiably so a hassle. But that ignores people that dropped the bevy game engine due to the difficulty to learn it.

I have struggled understanding some core concepts because the doc has, at times, 5000 less character than it should have considering how important the underlying tool or concept is.

The argument that anyone interested in AI generated doc should do their own has one big flaw:

Knowing how to use AI to make good things is a skill to learn, that not everyone has and far less people are interested into.
I read often that anti-AI people just don't know how to use. I cannot asses how true this is, but it may be, what do I know?  I currently assume this to be valid.

I never said that generated doc is better, but to let the AI friendly community making friendly documentation without shaming it, being angry, and aggressive each time it is publicly mentioned. More often no concrete evidence the content is harmful is ever clearly pointed, it's all about general moral concerns.

I don't know if it was AI generated but https://github.com/pavlov-net/claude-plugins/blob/main/bevy/skills/bevy/SKILL.md has been the best documentation on bevy I found so far, tainted coders one is good too, but not up to date enough (and that's fine it's free)

Reviewer 30: >Knowing how to use AI to make good things is a skill to learn, that not everyone has and far less people are interested into.
I read often that anti-AI people just don't know how to use. I cannot asses how true this is, but it may be, what do I know? I currently assume this to be valid.

Can we please avoid discussion that amounts to "you are using the tooling incorrectly"/"you are anti-AI and don't know how to use tooling"? It's very frustrating to read (especially when skeptical folks are already being *incredibly* gracious) and I don't find it is particularly productive. Do you have any concrete arguments against the content of what I said? How might you solve for the real issues I brought up when using LLM generated prose?

Reviewer 23: > Can we please avoid discussion that amounts to "you are using the tooling incorrectly"/"you are anti-AI and don't know how to use tooling"

First if you read what I said I clearly stated this was a possibility which I assumed to be correct for the following of my argumentation.
This implies that this could be totally wrong and if so the following would be on frail basis.

I am not using AI, I have no particular knowledge on how capable or incapable it gets. I'm just chatting with people that do use AI and get to trust them when they say something about it.
When I'm mentioning 

> AI generated external documentation

I'm specifically talking about a community driven AI generated external documentation that anyone could contribute to correct humanely in case some wording is in need of re-wording and some facts to be refactualized.
Keyword: external. This does not impact any internal documentation bevy like the current ai policy rules out  but it does concern the community.
Current community that would rather see people not having documentation, judgement likely based on their own familiarity with the project, than having something generated with a LLM.

Reviewer 27: > I suspect a UI for LLM generating docs lazily/on the fly as you navigate through them either already exists or will show up soon.

DeepWiki is basically this. Note: It has pre-henerated pages, but it is really designed to generate docs interactively in response to a prompt.

Reviewer 9: I know I have argued about this before, but I feel like the policy is still ambiguous about tabular data, in that it could be interpreted as a violation or not depending on who's doing the reading. So apologies if this seems like I am being repetitive.

One of the most valuable kinds of LLM-generated criticism is using the tool _as a linter_: coming up with a lengthy and detailed report of all the problems it found. This is particularly good for me because while I tend to be good at spotting big picture, architectural issues, I sometimes miss small details. At the same time, I am free to exercise my judgment - I don't always agree with the machine's recommendation, just the same as I would with a more primitive linter.

While there's no issue with me using this privately, for large projects there's a desire to be able to divide up the work amongst multiple people, which necessarily requires sharing the list of errors. While I have in the past "lightly edited" such lists (usually by deleting the introductory paragraphs, or places where it explains too much), it's (in my opinion) unreasonable to ask me to re-write a list of 50 syntax errors "in my own words".

Nor is there any problem about not having a person behind the text, because each lint item still has to be acted on and evaluated by a human who presumably understands the impact of the change.

I would be fine with a requirement to disclose the fact that the list was AI generated, or even put it in a collapsible section so that people who don't want to read it don't have to. I do _not_ think it would be acceptable to post such a bare list as the only content of a GH issue, without any human-authored preface, and in fact I would consider that extremely rude.

>   Publicly shared design documents (or problem definitions) must be personally authored, synthesizing your own understanding of the problem space and any proposed solution.

Author: This was also contentious: some AI-users within our community find that a generate-then-edit workflow to be more natural and effective than "redraft from scratch".

Others (e.g. @cart and myself), wanted to keep this, willing to sacrifice speed and convenience in favor of forcing folks to rethink their plan and use a more natural, concise voice.

Reviewer 17: > 7. **Design:** Writing is thinking.

I would be in favor of dropping/changing this part -- "Writing is thinking" takes me back to to my early school years when I realized I learned differently than most kids and that there are lots of ways people process, learn, absorb, and retain information -- and this kind of feels like it is trying to pick one way for everyone.

We've already stated above: "Contributors are wholly responsible for their own submissions, regardless of how they are created.  You must personally understand every line of code and documentation that you submit.  Contributions must be original (or appropriately credited and license-compliant), well-made and crafted with care: something that you are proud of."

To me that covers "don't post overly verbose things you don't understand."

Reviewer 13: I am against dropping this part. 

Broadly speaking, if someone can't explain things in their own wording then how can anyone else be sure they understand. While I'm not against people using AI to explore a design space (as much as I'll avoid that personally), I would prefer being able to know they at least had to sit down and actually think about it, and try to understand it. 

Furthermore, if we allow people to just use LLMs to fill in this space it increases the liklihood that people will just fall into cognitive surrender/offloading which would fundamentally increase the workload on maintainers.

Reviewer 18: In the public-facing part of this, I mentioned feeling that it is a "bare minimum" restriction; I feel only slightly less strongly about this for "internal" design documents. Even if you are in favor of LLM usage, it is likely going to be beneficial to do another pass redrafting from scratch for the same reason we do drafts at all; multiple iterations almost always shows you flaws to be fixed and improvements to be made. This is likely especially true when the original draft was generated by an unreliable program in the first place. For that reason alone I think this restriction should be kept.

Reviewer 16: I agree with this rule a lot. I use AI to help synthesize design documents a lot and help _me_ understand, but I almost universally find them terrible to read. It tends to be a wall of text with a lot of implicit assumptions, poor pacing, etc.

I agree that having a human re-write things to ensure that _they've_ understood it and put thought into how _others_ should understand it is valuable.

If models improve, we can always re-evaluate this document.

Reviewer 5: > I would be in favor of dropping/changing this part -- "Writing is thinking" takes me back to to my early school years when I realized I learned differently than most kids and that there are lots of ways people process, learn, absorb, and retain information -- and this kind of feels like it is trying to pick one way for everyone.

We could always rephrase it as _Writing is one of the main ways humans think_ or use some more neutral language

Reviewer 34: For me this isn't about output quality. It is about forcing a human brain as a bottleneck into the design workflow. It ensures our hands are always on the wheel, and that authors are actually internalizing their proposals at a fundamental level.

Reviewer 17: > For me this isn't about output quality. It is about forcing a human brain as a bottleneck into the design workflow. It ensures our hands are always on the wheel, and that authors are actually internalizing their proposals at a fundamental level.

The intent here makes sense, but I still think it forces people to work in one particular way that works well for some, but not all.

I'd also propose just removing it to actually see if it is a problem rather than banning to start.  I don't think it will be, but if it is, add it then.  Otherwise you have people potentially hiding things...

Reviewer 9: > It is about forcing a human brain as a bottleneck into the design workflow.

The problem I have with this argument is that there are some kinds of workflows that humans are traditionally very bad at; in those cases, the hands may be on the wheel but the captain is asleep on duty. Change-blindness, repetition-blindness, just simple boredom can lead to mistakes.

Human brains are storytelling engines, and what they are good at is creative narration. What they are not good at is consistent repetitive ritual. But bug reports and design documents do sometimes need this kind of text - tabular data extracted from other sources and summarized. This is the kind of thing that would normally go in an appendix, but modern bug tracking systems don't have this concept.

Reviewer 17: [Am going to post this on both proposed Rule 6 and 7 since they're tied together]

I think both of the writing rules (6 & 7) should both be changed.

They both fall in to the "things we can't detect" trap and I expect we'll have the same behavior other parts of this are trying to solve for with code (or people just doing it and not mentioning it).  Rule 1 already covers "You must personally understand every line of code and documentation that you submit."

I would keep the values behind them, like human voice, real understanding, and careful editing, but change the wording to an expectation instead of a ban. 

On rule 6 (verbosity): I agree that a lot of AI writing today is long and hard to read. But I think the rule is aimed at the style of the writing rather than the real problem, which is writing that is low quality or that the author does not understand. On both ends of it, people can write lots of long unclear text, and when done well edited AI assisted text can be great.

Suggestion: "Writing: Public-facing prose, such as module docs, release notes, long PR descriptions, and book chapters, should be clear, concise, and written in your own voice. You may use AI to help, but the final text should be edited such that it is genuinely yours and something you are comfortable putting your name on. We would rather read something short that you wrote yourself than a long draft you did not really write; our editors can help expand and revise it later. Using AI as a critic, or for small mechanical fixes like grammar, is fine."

Rule 7 ("writing is thinking"): People learn and process information in different ways. Writing things out by hand is how some people think, but it is not how everyone thinks (my middle/high school teachers thinking this was the best way for me to learn did not go well...). A rule that requires everyone to personally author their design docs takes one way of working and applies it to everyone, and it still doesn't guarantee understanding -- I've spent half my career sitting in presentations where people clearly had no idea what they were talking about.

Suggestion: "Design: When you share a design document or problem definition publicly, it should reflect your own understanding of the problem and any proposed solution. AI can help you explore options, but the write-up you share should be something you can fully explain and defend on your own. Working through a design yourself takes more time, but it usually helps you understand it better and often leads to a stronger result."

Reviewer 24: Strongly against removing this. It is vital that, at a minimum, design documents and similar writings be human-generated. LLMs excel at generating things that are incorrect but, at a surface, _appear_ to be correct. Much as I dislike their use in code, it is still possible to run tests to ensure the behaviour is correct. In a design document, this is not possible.

Reviewer 13: The whole "Forcing someone to work in a way that might not work well for them" argument doesn't really hold water for me because, fundamentally, *all workflows might cause that outcome*. By contributing to Bevy you inherently have to use the workflow that Bevy uses (to an extent). By being on Github we force a particular Workflow.

Fundamentally all projects that allow someone else to come by and contribute require that person to, at least to some extent, adopt and use the Workflow that project dictates.

As for the "Things we can't detect trap", there is already plenty in this document covering things we can't necessarily detect. At the same time, however, we must necessarily accept that this document is not 100% about enforceability, but also about setting how the project works and how the community should operate.

Reviewer 12: Design docs are a demonstration of understanding, and for that, *you must do the work*. You cannot sidestep it as you are translating one's thoughts and ideas into a plan that others can read and critique. Putting this down as a requirement isn't going to prevent folks from generating a doc; what it describes is the path of success because you writing a design doc down yourself will make it easier for you to defend and explain it. People will ask questions about details and decisions made, and if you *did the work*, then answering these questions will be an easier time.

If you did not do the work, then it will be much harder to answer them. And as a result, much more likely the design document will not be accepted.

> 8. **Art:** Out of respect for the craft of artists, AI must not be used for creating or manipulating art assets: find a permissively licensed asset, commission something appropriate, or make one by hand.

Author: **THREAD ON AI IMPACTS ON ART**

Reviewer 18: I have concerns about bevy's ability to attract artists to the community in the long term with an at-all permissive license towards AI. I haven't personally seen the dynamics play out of how artists respond to programmers' usage of LLMs very much; but it is clear that artists are **extremely** anti-AI, and it seems likely to me for this to extend into a field that isn't necessarily their specialty e.g. programming. We currently have very few non-programmer artists in the community due to lack of tooling, but I would be concerned about a permissive AI policy (even one that bans the usage of generating "art" like this one) leading to a general distaste/rejection of Bevy as a project by artists.

Godot Engine currently seems to be aiming at a slightly more restrictive policy wrt AI, and is also open source, more popular, and has tooling already. I would be concerned, in the future, to see Bevy lose access to artists who would rather stick to Godot for the reasons mentioned above. Popularity and tooling will be fixed with time but an AI policy won't necessarily be.

Reviewer 21: This kind of suggests that coding/writing is not an art, or is somehow lesser so it doesn't get the same respect.

Reviewer 30: I agree with the sentiment on it being a bit confusing. The real reason to my understanding is that both can be considered art, but the visual arts space is much more reactive to gen AI use, likely because they have had a much more severe change in their material conditions as the result of models stealing artwork for training data. 
Open source is far more corporate friendly, and companies have created a culture where programmers feel less ownership over what they create, which I think leads to a more accepting view of AI. I'm not really sure how to communicate this in a way that is *logical* though...

Author: As a programmer, I am comfortable speaking directly about our values around code, and what is permissible. I am *much* less comfortable doing so for visual artists and musicians (I am but a dabbler with terrible fine motor skills), especially in the face of much more negative sentiment. If we develop a community of artists who want this subsection of the policy changed, I'm happy to follow their lead!

> This kind of suggests that coding/writing is not an art, or is somehow lesser so it doesn't get the same respect.

I see where you're coming from here :) Suggestions that improve on this without exploding word count would be welcome.

Reviewer 16: I have mixed feelings on this, personally. I made bevy_solari in large part because I want game development to be accessible. A little bit of a contradiction, right? Pathtracing is extremely intensive, and you need a fairly good/expensive GPU to use it. But pathtracing is also drop-in: you don't need to understand shadow mapping, direct vs indirect light, lightmaps vs irradiance volumes vs screen space lighting, etc. There's just one "lighting behaves how you expect it" button, and I think that's really valuable for beginners. 

Generating high quality materials is an extremely difficult, expensive, and time consuming task, even for experienced artists and game studios, which leaves beginner 3d games tending to look pretty awful https://github.com/JMS55/RainbowBall, even with perfect lighting handled for you. Papers using generative models https://research.nvidia.com/labs/rtr/publication/xue2026videoneumat/ have a lot of opportunity to help with that. I hope that it raises the floor of things without devaluing the ceiling of what actual professionals can achieve with time and care, but yes, the unfortunate reality is that people often will often see "good enough" and leave it at that.

Reviewer 5: I agree each _trade_ should have the freedom to define its own ethics, so if musical/visual artists are not okay with AI, but programmers are, that is a meaningful/valid place we can arrive at. This practice even already de-facto exists (e.g. sampling, musical covers, parody, fan-games, etc)

I agree though that framing it naively as art vs non-art opens up a can of worms around "is code art?" etc.

Let's be more specific with the definition (e.g. “visual, music, sound assets”) and instead of respect for the craft itself, let's maybe say something like “respect for the broad ethical norms emerging in the field of ..."?

***EDIT:*** Oh another important point: Should this rule also apply downstream to things like:

- Bevy game jams
- Other ways in which the Bevy foundation or community may feature projects
- General sharing of pieces in Discord, etc?

Reviewer 30: I think what makes this difficult is that the driver of this disparity community attitudes. If we don't permit AI art, it's because *we've decided we just don't like it*. I can't think of a cohesive argument that bans AI art while allowing AI generated code.  The best argument I can think up is that AI art harms artists disproportionately, but that's not particularly watertight.

Reviewer 17: > 8. **Art:**

I think this should probably just be removed, mostly because I'm not sure it actually buys much besides arguments over "what is art."  Given the nature of Bevy it doesn't seem like we're suddenly going to be over run with assets, nor do I see reason to dump a bunch of AI generated images on the website, but to @JMS55's point if it does help with testing & development that seems like a benefit.

Seems like basically banning likely has little impact, but leaving it out means maybe some people finding limited/specific uses that are helpful and productive.

Reviewer 35: > This kind of suggests that coding/writing is not an art, or is somehow lesser so it doesn't get the same respect.

I'm fine with having a relatively expansive understanding of what counts as art / art practice, but I think it's really important to note in these conversations the relative precarity of someone who works as a software engineer versus a visual artist.

Reviewer 13: I'm against dropping this for the main fact that I keep bringing up in these discussions:

**Irrespective of all else, Bevy wishes to attract artists. The artists we wish to attract are largely anti-AI. This policy is already likely to cause some malcontent with those individuals. This mitigates things a bit and likely would make it a bit easier for an artist to be willing to even consider using Bevy.**

This prohibition sends a signal that Bevy is not pro-AI as a project, but is merely adopting a more pragmatic approach. 

Further, I'm certain that the Foundation and moderation are more than able, and willing, to enforce this in a way that makes sense. Given that this policy only applies to Foundation controlled spaces (so broadly, the Org, This repo, and everything else) I fail to see how this ban would meaningfully cause issues if someone *does* use AI to generate assets or art for their own projects. It just says the Foundation and its projects will not do so.

Reviewer 27: > This kind of suggests that coding/writing is not an art, or is somehow lesser so it doesn't get the same respect.

I personally do believe that coding is proportionately less *creative* than art/writing (as it contains an element of engineering that these disciplines typically don't). And that AI usage is much more problematic in creative domains.

Reviewer 5: > If we don't permit AI art, it's because we've decided we just don't like it. I can't think of a cohesive argument that bans AI art while allowing AI generated code.

One angle is that on the subjective ↔ objective spectrum, what we consider to be “art” sits much closer to the subjective side, and what we consider to be “coding” much closer to the objective side. Of course there are subjective elements to “coding”, and objective elements to “art”, but it's much easier to measure the objective efficacy of a piece of “code” (benchmarks, unit tests, type checks, lint rules, algorithmic complexity, etc) than of a piece of “art”. (palette, composition, aesthetics, proportion, technique, medium, emotion, etc) Sometimes in code there's only one solution to a problem, and when there's more than one, they generally sit on a finite, well-defined trade-off curve. Not the case with art.

The other angle is purely based on/ social norms. If one camp feels largely negatively about using AI, and the other feels largely positively about using AI, honoring each space's preference contextually is one of the few ways to reconcile this productively.

> 11. **Translation:** LLMs may be used for translation, but you must provide the original untranslated text as wel

Author: This is more severe than my initial draft. Providing the original untranslated text was framed as optional in that case. Others expressed concern that "oh yeah I'm just using LLMs for translation" is a common get-out-of-jail-free card for reckless slop, and wanted to close that loop hole.

I am particularly interested in feedback from non-native English speakers here: is this acceptable? Is it too onerous?

Reviewer 13: While I'm not a non-native English speaker I have been a part of many communities where I've been a non-native speaker trying to discuss with people are largely are. 

Generally speaking, at times including the original text was somewhat helpful as there may be nuances or things that I didn't fully understand or grasp when trying to speak the language (or at times having a machine translate the sentences). While this was done for the purposes of language learning, I think that it provides a similar benefit. Additionally, it has the side effect of allowing someone who is fluent (or at least of a high level) in both that language and English be able to potentially better explain other issues and vice-versa, and if someone wants to use a different tool for whatever reason to translate the text or to make it easier to understand the original can be used for that as well.

Reviewer 5: > "oh yeah I'm just using LLMs for translation" is a common get-out-of-jail-free card for reckless slop, and wanted to close that loop hole

I don't expect this to be an issue in practice, LLM translated content vs LLM authored content already looks distinct enough that we'd be able to tell by the vibe alone, but I don't mind requiring including the original. I just don't know how useful it will be considering only a fraction of contributors (possibly not any of the relevant reviewers/SMEs) might understand the original language.

Reviewer 30: IME I find LLM translation is superior to traditional translation methods, so I would advocate a bar on machine translation in general if LLMs are not allowed for translation.

Reviewer 36: What if translations are AI assisted but manually revised? In such cases, there is actually no original untranslated text.

Author: > What if translations are AI assisted but manually revised? In such cases, there is actually no original untranslated text.

Hmm! Personally I think I'm fine with them being posted without the original text and an optional disclosure: reaching natural tone / fluency resolves my concerns around personal touch and bad faith "oh it was just a translation" that makes moderation challenging.

@james7132, this was your suggestion, so I would appreciate your thoughts on this.

Reviewer 13: I would personally prefer the original be provided regardless, as it still has utility IMO

> 1.  **Good faith:** Welcome others to the project and to the community, and presume that they are contributing in good faith.

Author: ```suggestion
12.  **Good faith:** Welcome others to the project and to the community, and presume that they are contributing in good faith.
```

> 12.  **Good faith:** Welcome others to the project and to the community, and presume that they are contributing in good faith.

Author: This was inspired by GCC's AI Policy: https://forge.sourceware.org/redi/gcc-wwwdocs/commit/4d0793a6a14bf9bfe9e92ac1599840780355199d

Thanks!

> 

Author: **THREAD ON ENVIRONMENTAL CONCERNS**

Reviewer 5: Water usage is blown vastly out of proportion and unrealistic. Other environmental concerns (e.g. pollution, noise) and also infrastructure and societal concerns (energy, _municipal water_ supply, zoning) are valid. These concerns already exist with other project-related activities which are not AI-related (e.g. if we do in-person talks / conferences, and people travel by plane).

Should we take on an official position of:

- Neutrality?
- Harm reduction?
- Advocacy?

There are various degrees to which we could do this, if we choose to. We could favor open/local models, could dedicate a part of the foundation budget to carbon offsetting/advocacy, could on a more extreme case even _require_ proof of carbon offsetting when accepting LLM contributions? (Probably hard to enforce)

> 

Author: **THREAD ON LABOR CONCERNS**

Reviewer 12: I'm just going to put down one of the big reasons why I am very Anti-AI, and it has been how LLMs and generative AI has been used to AI-wash productivity mandates, lay-offs, suppress wages, and otherwise instigate anti-labour actions. Having survived a lay-off recently and been put under an AI-washed productivity mandate, arguments on the utility of LLMs basically sound to me like "it's okay to have all these bad things happening to other people, because _I_ find some personal benefit!". Or worse, "it's just a tool!" completely ignores the externalities that it has been creating even for software engineers, let alone the public at large.

Companies have been getting away with the exploitation of mental labour, and then *selling that back* to people, and we want to allow/invite that in to *community* spaces? It doesn't matter even if you are using local models instead of the big tech ones, because they are all being created from the same infra, distilled from frontier models created with billions in money and all dependent on scraping vast amounts of data that they are likely not actually privy to! People have been imprisoned for doing what these companies are doing, but nothing is done while billions are at stake.

"It's just a tool" is hypocrisy, if one cares about community health and the people within them. It cannot be separated from the impacts it is having, and to pretend otherwise basically undermines the supposed values this community has.

Reviewer 5: @Bluefinger First of all, lots of respect and empathy for what you are going through, also having been through repeated rounds of layoffs and reorgs at various previous workplaces, and having friends in similar situations to yours right now, I know how much this sucks and takes a huge toll on mental health and the ability to enjoy work and life in general.

I agree "just a tool" is a bit obtuse. For example, weapons are also technically just tools, yet they are a highly contentious and consequential socio-political issue.

That said, disallowing AI here has failed to move the needle towards decreasing AI negative externalities in general, and in fact it's directly causing those negative externalities to leak into our community by increasing moderation burden, creating suspicion and animosity, and incentivizing dishonesty from bad actors.

One thing we can confidently do as a non-profit/volunteer project is to explicitly never _require_ AI use, and always welcome and cherish human-only contributions, something that a company that has the fiduciary duty to maximize shareholder profits and strict authority/control over its employees can't do.

Reviewer 12: > That said, disallowing AI here has failed to move the needle towards decreasing AI negative externalities in general, and in fact it's directly causing those negative externalities to leak into our community by increasing moderation burden, creating suspicion and animosity, and incentivizing dishonesty from bad actors.

Please do better than inevitablism arguments. "Bad actors making things painful, therefore we should cede ground" is an awful argument to make, and what it actually highlights is that there isn't the support the community needs to hold firm on the values it has signposted. We don't need to move the needle externally. We just need to be a space where people can be comfortable and enjoy being in. The fact that there are stressors being inflicted on us doesn't mean we should dilute our values to be rid of them.

Part of the stressor _is_ the incessant encroachment of LLMs, the fact it is being forced, the fact that "No" and consent is being ignored.

We are living through times where rights are being actively eroded. And you want to _cede ground_?

Reviewer 5: > We just need to be a space where people can be comfortable and enjoy being in

Which was not being the case because of the current AI policy, as it was creating a climate of constant suspicion, animosity, and undue moderation burden

> We are living through times where rights are being actively eroded. And you want to _cede ground_?

I want to choose my battles and spend my energy on things that I believe will be fruitful. In my view, taking a strong ideological stance of standing ground against AI is highly unproductive, so I'd rather focus on something else and hopefully just have fun building stuff in my free time.

There's an nearly endless supply of injustice, oppression, and rights erosion to fight against in our lives, one has to constantly chose to cede ground on some things, to focus on others, as it would be impossible to fight every battle, stand for every noble cause, and avoid all forms of harm.

The way I see it, and perhaps I'd like things to be different, if my whole life's work and intellectual output has been put in a blender along with that of the rest of humanity, and instrumentalized to build a machine that can do useful work, I might as well benefit from it in some way that I find amenable and non-harmful, and not spend time policing what others do with it.

I don't see this as defeatism/inevitablism, but rather as pragmatism.

Reviewer 12: > I don't see this as defeatism/inevitablism, but rather as pragmatism.

I read it as that, and because of that, I suggest we don't continue with this discussion because we are ideologically opposed.

Reviewer 33: > Which was not being the case because of the current AI policy, as it was creating a climate of constant suspicion, animosity, and undue moderation burden

I don't think I fully understand how the revised policy solves this problem. Isn't there a risk that it just shifts the moderation burden from "try to detect use" to "try to detect non-disclosure"? It's just not clear to me how a community can effectively incentivize disclosure when non-disclosure is not reliably detectable and (if undetected) carries zero costs (to reputations and relationships).

The *individual* incentives still seem skewed towards non-compliance under this new policy. But this policy still needs to be enforced by moderators. That's basically the same tension driving the current "climate of suspicion, animosity, and undue moderation burden". From this perspective, the new policy feels a little like a "worst of both worlds" option: it invites AI-authored content and all the attendant negatives, while failing to resolve the current problems. Some amount of negative emotions might be an unavoidable consequence of enforcing any restrictive policy. 

On the flip side, I do see the value in creating (or striving to create) a culture of honesty. Maybe strong incentives can be created at the community level through positive interactions over time. Even if incentives do remain misaligned, it's possible that these relaxed rules end up, on balance, creating less negative emotions and experiences for maintainers and contributors, and increasing honesty. I actually feel optimistic that the Bevy community could pull that off, based on what I've experienced in the brief time I've been here. I do think there is a risk, though, that that doesn't happen, and that this new policy fails to resolve existing compliance problems while creating entirely new categories of problems.

Reviewer 30: > Bad actors making things painful, therefore we should cede ground

I don't believe this is the strategy. In my view, bad actors are *already making things painful* and LLM code is *already in the repository, undeclared*. This is mitigation against damage already done, not an invitation for more AI contributions, which is why, e.g., the general consensus seems to be we don't want to add an AGENTS.md.  Inevitability to me is limited to "AI is a necessary tool because of the productivity gains/it is industry norm", which most of us agree to be false. It is just fact that *many* people are using LLM tooling and ignoring that fact when creating policy leads to blind spots.

The outcomes to this seem to be to either invest significant moderation time hunting for AI generated code, at cost to community health (to which Alice posted a link on how that is a pretty fruitless endeavor), or move towards disclosure so more informed decisions can be made. I'd still be happy for a complete and outright ban, but I understand the downsides.

> 

Author: **THREAD ON BEVY COMMUNITY POLITICS AND VALUES**

Reviewer 18: My own comments here are as much driven by politics as by practicality. I originally joined bevy out of only mild curiosity and became far more invested in the project as I realized how seriously the project takes politics and how well-considered those politics are; making bevy very explicitly a space for queer creators. and not one for trolls, bigots, or anyone else who will seek to push out others on the basis of their identity. 

I was equally captivated by the bevy community's technical prowess and community management; processes like allowing anyone to review contributions, and encouraging this as an onboarding tool, was entirely new to me and showed me how cutting-edge bevy is when it comes to managing an open source community towards a common goal. It was deeply impressive and has left a permanent mark on how I perceive and work with other open source projects. 

Which is what makes anything less than a wholesale rejection of AI still so disappointing. As mentioned, there is no silver bullet to block all AI contributions; given a motivated enough actor, it's not hard to get a change merged that technically uses AI in some sense. It is disappointing to me to not try and block it anyway, to say "we see the harm these technologies are causing, we do not want a part of it, we will not make space for it". On a technical level the use of AI has still not amounted to anything outside the brackets of "barely functional slop" and "not worse than before". On a political one there's too many issues to reasonably list or explain them all. 

A policy that is not zero tolerance on AI usage, regardless of how well intended or designed still fundamentally feels like compromising on the values of the community. 

This is not exactly an argument to "just say no"; I know that that will not be implemented, not right now at least. Rather I am trying to explain my own feelings on this, which I think a not insignificant amount of the community will share.

Reviewer 5: Both the Bevy project and the community have always struck me as highly pragmatic, so I think this policy is overall a good fit.

---

> On a technical level the use of AI has still not amounted to anything outside the brackets of "barely functional slop" and "not worse than before".

Hard disagree on that one. It was perhaps arguably the case roughly 6-months to one year ago before context windows, harnesses and RLVR got good enough, but the capabilities and use-cases have moved a whole lot. There are now entire class of tasks that would be impossible or at least uneconomical that are now possible with the help of LLMs.

Browsers have patched hundreds of exploits (in human-written code) that would otherwise not have been found. People have ported million-line codebases between languages in a week with only very minor regressions.

There's a very broad discrepancy between what's easily available to the general public (e.g. in free or even $20 subscriptions) versus what is now available at the top tiers of budget/spend (e.g. $200 or more) if you know what you're doing. 

It's also unfortunately easier than ever to slop-cannon / one-shot crappy demo projects that rank well on Twitter, so this might bias your perception to the lower end of the spectrum.

Reviewer 24: I concur in a disappointment Promethia has expressed that this is not a full rejection of all generative AI usage. Given at this point such a policy seems unlikely though, I would suggest that the Bevy _project_ has a strong stance against the use of generative AI, but for _practical_ reasons, uses this policy (under the argument that a broad ban would encourage LLM users to hide the tooling usage, and the aforementioned arguments about a lack of trust).

Reviewer 5: > I would suggest that the Bevy project has a strong stance against the use of generative AI

Do we broadly have such a strong stance, though? I feel like as a community we're fairly heterogeneous in our opinions on this, and any such clear position could only be achieved right now via top-down rule rather than any sort of consensus or vote, and it would be far more harmful to the community at large than a compromise solution.

Also the distinction between generative and non-generative AI isn't a very good line in the sand, for example, is the Bevy project against DLSS?

Author: We do not, for the record, have consensus on the politics, ethics or appropriate use of AI, at any level: maintainers, core contributors, contributors broadly, community members or Bevy users.

Many of our most *vocal* community members are strongly opposed to AI, but having talked to dozens of people at length about this topic in public and private that appears to mostly be a selection effect. Strong opposition to AI is a sizeable minority view, with mixed or moderate feelings as the plurality, and clearly pro-AI viewpoints as the smallest group.

Despite that lack of consensus, we must still find a policy by which to run the Bevy project: a pragmatic approach with meaningful compromises on both sides that centered values and goals *for Bevy* was the one that I felt was most appropriate, and the one that I personally believe in.

Reviewer 24: > Do we broadly have such a strong stance, though?

I do not have polling data for the community at large. From a very quick skim of discussion on this repository, I see more agreement (including reactions) with the strongly anti-genAI position than even being pro-genAI to even a small degree, but that may be wrong. I am advocating for the position that I hold, as we all are in any discussion.

> Also the distinction between generative and non-generative AI isn't a very good line in the sand, for example, is the Bevy project against DLSS?

The distinction between generative and non-generative AI is something we must draw, as the term "AI" on its own is utterly useless. Pathfinding is a kind of AI, computer vision is a form of AI, and large language models are a form of AI. These three things are categorically different, and most all opposition exists to the third category. As for DLSS, I'm not entirely sure what you mean—that's something managed on the GPU level rather than anything Bevy would be involved in, AFAIK.

I would of course be interested in seeing community polling on the opinions of the Bevy community at large,

Reviewer 5: This is a somewhat pedantic reply, so I apologize for it in advance. 🥲 I've seen the generative vs non-generative AI point brought up many times at various spaces, and it is a pet peeve of mine.

Generative versus non-generative only refers to whether the AI model is doing classification of existing data or generation of new data based on a prediction. For example, various time-series forecasting techniques and Markov chains _are_ generative AI.

You can often run a classification model "backwards" to do generation, and this is largely how diffusion image gen works: Start with a random noise image, a classification model and a label (e.g. "fish in a bowl"). The classification model measures how well the image fits the label. (Initially very close to 0, since it is noise) Then produce perturbations on the noise and hill climb towards maximizing the label to 1. Do this with neural networks from the 1980s for number/face/shape recognition and it will largely “work”. Recent breakthroughs are mostly in data and compute scale, and a few tweaks to architecture, but not a drastic change in principle.

Pathfinding and computer vision are applications/fields of AI research, not categories or separate types of AI. Modern multi-modal LLMs _are_ highly powerful computer vision systems, and you can even use them for various classification tasks, either via constraining the generation or directly reading the logits from the neural network.

None of this is nebulous/inscrutable/inaccessible, it's just math you can learn and implement in a weekend. And arguing that some kinds of math are morally objectionable while some are not makes no sense. So the key moral objections must be not due to the underlying AI technique being generative vs non-generative, but rather: 

- How training data is sourced (e.g. copyright infringement, lack of consent, under-compensated labor)
- What the technique is used for (e.g. making hiring decisions, surveillance, or automating something held deeply as intrinsically human, such as art, thinking, socialization)
- Direct or indirect consequences or harms from use (e.g. job displacement, dependency, reduction of skills, psychosis, misinformation, environmental damage, etc)

Re: DLSS, it _is_ generative AI, and we had to add explicit hooks for it in the engine: https://github.com/bevyengine/bevy/blob/main/crates/bevy_anti_alias/src/dlss/mod.rs So claiming we're downright against generative AI would make no sense.

> To help you understand where Bevy draws the line, the following AI-related behaviors are not acceptable:

Author: These were drawn from real problems that have been experienced within the Bevy community. Community feedback on other behavior that they have questions about, or find intolerable would be particularly welcome here <3

> 

Author: **THREAD ON BANNING AI-SUBMITTED CONTENT FROM BEVY COMMUNITY SPACES**

Reviewer 30: These rules make sense to me. One thing I have considered is instead labelling my own shared content as *not* using AI, instead of relying on the inverse, since I personally am more invested in accurately labelling my content. However I can also see how that can lead to a less-than-ideal community environment.

Author: I quite like self-labelling as AI-free: I think that this is a healthier statement about personal values. Mirrors the "all organic" labels, and lets people choose projects to support that align with the things they care about about.

Reviewer 18: Definitely in favor of this (although curious if it technically fits in this proposal, or should be placed elsewhere). Generated posts/memes/images/videos are, at best, a mild amusement, and more likely low quality slop that lowers the quality of the community space. Also generally related to troll/bot activity, based on what I've seen on other platforms (e.g. reddit).

Author: On a similar note: we will need to decide a policy for game jams. The models are *far* too good at generating low-quality code quickly, and I worry that a "AI for code is allowed" solution like we've done previously will result in an unfair advantage  / domination in impossible-to-moderate but bad vibes way.

Current thinking is to have two categories: one fully restricted (only libraries, hand-authored code, no external assets) and a second that's open (any asset, LLM-assisted code allowed). This is not a values statement, just a pressure relief valve to discourage cheating.

IDK! Hard decision here. I'd rather stop running game jams than spend hours trying to adjudicate if folks generated assets or code for them though 🙃

Reviewer 30: I'd personally prefer we don't open a space specifically for AI contributed games on itch. I think the concerns around AI sentiment in the indie space could hold firmer on itch, and allowing for a genAI category in a game jam might not be good PR, (and id also prefer we just not spend resources on something like that - i'd much prefer we encourage human use of bevy whenever possible because of the higher return on investment). 

I think we can direct users to fill out the generative AI question on Itch as accurately as possible and let users decide, or completely ban AI contributions.

Reviewer 16: I've never loved the idea of game jams/hackatons to begin with, but my experience with AI in hackathons is that the winner makes something impressive, but has almost no understanding of what they've made (on a technical level), and it ends up being very fragile because speed and general vibe is prioritized above _all else_.

Game jams with AI could be fun for exploring game _design_, but as a dogfood of the engine, it'll be useless.

I don't see how we would outright ban AI though, as it would lead to the same problems this document discusses.

Reviewer 5: > but as a dogfood of the engine, it'll be useless.

Not necessarily, it could be useful for stress testing and finding bugs, for example

Reviewer 16: I don't know that I have strong conviction that an AI would recognize that something is a bug in the engine, vs thinking that "oh, I generated some wrong code, let me try again". But potentially.

Reviewer 5: I have personally witnessed this first-hand a few times, I am confident at least Fable/Kimi K3/Sol-class models can do it

Reviewer 17: > I have personally witnessed this first-hand a few times, I am confident at least Fable/Kimi K3/Sol-class models can do it

Agreed, I've also found that once I had some local engine changes, it was much happier to go fix other engine bugs

Reviewer 32: There was a "debate" about AI use in the in-person jams that are ran in my local game dev group, and ultimately it resulted in a full ban of AI. However, there was also a pretty good talk about why using AI is antithetical to game jams, which made people a lot more understanding of why AI was banned. If AI is banned from Bevy game jams (and I would support that), than I would also recommend having some info about why it's banned from the game jam outside of the engine policy.

Reviewer 17: > Current thinking is to have two categories

This seems like the right solution.  This lets everyone participate and show off their creativity which to me is the whole point of jams/hackathons/etc.

Ban one and you've put up barriers for what should be a low barrier to entry activity -- let people have fun, explore, try new things.

Reviewer 30: If I had to guess, game developers tend to feel a lot of artistic ownership over their projects, especially relative to something like an open source JS framework or command-line tool.  It is safe to say the majority sentiment among game devs is that games are a form of art, and they're often very personal and meaningful to the developer. 

I believe this is why the indie game scene has aligned itself towards the visual arts crowd when it comes to its disapproval of AI, whereas open source projects tend to be more permissive with AI use. 

I would push for a ban on non-insignificant LLM use in game jams, and a complete ban on AI generated assets (with a narrow definition - procgen should be allowed, given it is not trained on copyrighted or otherwise stolen material).

1. **It's just more interesting.** LLMs are impressive. They are far off from coming close to human creativity. I am not worried about LLM projects having an unfair advantage because I believe (at least as of now) the games they produce are wholly unremarkable.
2. **It's an opportunity to establish our community values**. I like this new AI policy because it is pragmatic; I'm also very disappointed that it's come to this. Keeping game jams human-focused is a place where we can follow that community spirit in a relatively undisruptive way, especially on Itch where even allowing AI submissions may limit participation.

I don't think this ban should include code created outside of the jam period, or for tooling - so users who have AI projects can copy their existing code, and tools like Jackdaw or Renzora would be allowed. Though would like to hear more opinions on that front.

Reviewer 13: I'm personally in support of limit/ban and explain why. If someone is found to submit AI-generated stuff regardless, they can be treated as anyone else acting in bad faith when it comes to the Jam rules.

Reviewer 5: > > Current thinking is to have two categories
>
> This seems like the right solution. This lets everyone participate and show off their creativity which to me is the whole point of jams/hackathons/etc.
>
> Ban one and you've put up barriers for what should be a low barrier to entry activity -- let people have fun, explore, try new things.

Yeah I think this is the most sensible approach. Kinda like car racing versus human racing, of course the cars are going to go hundreds of miles per hour, and humans won't, no reason to expect one to compete with each other. Some people enjoy both, some people enjoy only one. Everyone gets to participate, have fun and watch/play whatever they want.

Reviewer 37: Anecdotally I joined a game jam recently and used a LLM for code generation. The jam explicitly banned AI assets but allowed code. The code generation allowed me to solo dev some stuff that would've taken longer manually, it was flashy but ultimately uninspired, and I would've been better off partnering w/ ppl.

I felt a lot better once I ditched that and started working on another project w/ someone else. IMO using code generation kills the social aspect.

A large part of that for me in the moment was everyone around me asserting that they weren't using AI. That they were glad they weren't vibecoding. The itch submission page had a check mark to say that it was created without the use of AI. It wouldn't have been hard to not acknowledge I was using AI and I wasn't breaking any rules but it created a strong environment of "we don't do that here"

I'm sure that wouldn't generalize to a large open jam as well, but I'm strongly in favor of tagging projects as no-ai (and turns lies of omission into direct lies which might help discourage that behavior further)

> - discussing the politics and current events of AI

Author: Unlike some technical communities, Bevy has, for many years, permitted discussion of politics and current events in its community spaces. This is not / was not a carte blanche of course: appropriate spaces, not disruptive, be civil even when you disagree etc etc.

This is a non-trivial moderation burden but has, on the balance,  been worth it. "Politics" affects real people in important ways, and valuable interpersonal connections have been made. I even saw someone change their mind once!

> - criticizing products, companies, and public figures use of or stance on AI

Author: Note: we recently introduced a topic-agnostic rule against "dunking" in the #memes channel. This was a pre-existing concern (language and engine flamewars are an evergreen topic if unmoderated), but was also prompted by repeated low-effort dunks on AI.

Genuine critique, or assorted grumbling, is healthy and normal and serves an important social function (in moderation). The no-dunking rule was a continuation of our earlier "no politics in #memes"  rule, intended to both preserve the light-hearted nature of that space, and shift conversations about important topics towards healthier, less inflammatory and more thoughtful forms.

> - discussing best practices for using (or not using) AI

Author: Casual conversations about this should be quarantined to #machine-learning while on Discord, to allow folks who are sick of the topic to simply mute the channel and ignore it.

Reviewer 5: Fair, but maybe we should not phrase it as a _quarantine_ as that implies a strong prior judgement of value (e.g. something vicious/virulent/infectious/unhealthy)

Author: Mmm, I see your point. I would say #memes are quarantined though :p "Cordoned" is perhaps a more neutral term to use.
